<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Variable_Selection.utf8.md</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="customstyles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 66px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h2 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h3 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h4 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h5 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h6 {
  padding-top: 71px;
  margin-top: -71px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Data Analysis Exercise Website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./Continuous_Outcome_Analysis.html">Continuous Outcome Analysis</a>
    </li>
    <li>
      <a href="./Variable_Selection.html">Variable Selection Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="./author.html">About the Author</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="URL-TO-THIS-REPOSITORY-HERE">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<div id="overview" class="section level1">
<h1>Overview</h1>
<p>This document will guide you through some data analysis tasks with a focus on performing variable selection. For this exercise, we consider a categorical outcome.</p>
<p>While this is in some sense a stand-alone analysis, I assume that you have worked through the <em>Data Analysis</em> exercise and are familiar with the dataset and all the things we discovered during the cleaning process. We’ll use the same dataset here but focus on a different outcome. Other than that, the way to work through the exercise is like in the <em>Data Analysis</em> exercise, namely by writing/completing the missing code.</p>
</div>
<div id="project-setup" class="section level1">
<h1>Project setup</h1>
<p>We need a variety of different packages, which are loaded here. Install as needed. If you use others, load them here.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(<span class="st">&#39;tidyr&#39;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(<span class="st">&#39;dplyr&#39;</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">library</span>(<span class="st">&#39;forcats&#39;</span>)</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw">library</span>(<span class="st">&#39;ggplot2&#39;</span>)</a>
<a class="sourceLine" id="cb1-5" title="5"><span class="kw">library</span>(<span class="st">&#39;knitr&#39;</span>)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">library</span>(<span class="st">&#39;mlr&#39;</span>) <span class="co">#for model fitting.</span></a></code></pre></div>
<pre><code>## Loading required package: ParamHelpers</code></pre>
<pre><code>## 
## Attaching package: &#39;mlr&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     auc</code></pre>
<pre><code>## The following object is masked from &#39;package:caret&#39;:
## 
##     train</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">library</span>(<span class="st">&#39;parallelMap&#39;</span>) <span class="co">#for using multiple processors when running models through mlr</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">library</span>(<span class="st">&#39;visdat&#39;</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"><span class="kw">library</span>(<span class="st">&#39;gridExtra&#39;</span>)</a>
<a class="sourceLine" id="cb6-4" title="4"><span class="kw">library</span>(<span class="st">&#39;gtools&#39;</span>)</a>
<a class="sourceLine" id="cb6-5" title="5"><span class="kw">library</span>(<span class="st">&#39;pROC&#39;</span>)</a></code></pre></div>
<pre><code>## Type &#39;citation(&quot;pROC&quot;)&#39; for a citation.</code></pre>
<pre><code>## 
## Attaching package: &#39;pROC&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     cov, smooth, var</code></pre>
</div>
<div id="data-loading-and-cleaning" class="section level1">
<h1>Data loading and cleaning</h1>
<p>We will again use the Norovirus dataset.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">#Write code that loads the dataset </span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co">#You can of course re-use code you wrote in the other file.</span></a>
<a class="sourceLine" id="cb10-3" title="3">raw_data &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;norodata.csv&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-4" title="4"><span class="kw">head</span>(raw_data)</a></code></pre></div>
<pre><code>##   id  Author Pub_Year pubmedid EpiCurve TDComment AHComment      Trans1
## 1  2 Akihara     2005 15841336        Y                     Unspecified
## 2 17  Becker     2000 11071673        Y                       Foodborne
## 3 39  Boxman     2009 19205471        N                       Foodborne
## 4 40  Boxman     2009 19205471        N                       Foodborne
## 5 41  Boxman     2009 19205471        N                       Foodborne
## 6 42  Boxman     2009 19205471        N                       Foodborne
##   Trans1_O             Trans2 Trans2_O             Trans3 Trans3_O Risk1
## 1        0   (not applicable)        0   (not applicable)        0     0
## 2        0   Person to Person        0   (not applicable)        0   108
## 3        0   (not applicable)        0   (not applicable)        0   130
## 4        0   (not applicable)        0   (not applicable)        0     4
## 5        0   (not applicable)        0   (not applicable)        0    25
## 6        0   (not applicable)        0   (not applicable)        0     8
##   Risk2 RiskAll Cases1 Cases2 CasesAll     Rate1 Rate2   RateAll
## 1    NA       0     15     NA       15        NA    NA   0.00000
## 2    NA     108     43     22       65  39.81481    NA  39.81481
## 3    NA     130     27     NA       27  20.76923    NA  20.76923
## 4    NA       4      4     NA        4 100.00000    NA 100.00000
## 5    NA      25     15     NA       15  60.00000    NA  60.00000
## 6    NA       8      6     NA        6  75.00000    NA  75.00000
##   Hospitalizations Deaths   Vehicle_1        Veh1
## 1                0      0           0 Unspecified
## 2                0      0 Boxed Lunch         Yes
## 3                0      0           0 Unspecified
## 4                0      0           0 Unspecified
## 5                0      0           0 Unspecified
## 6                0      0           0 Unspecified
##                         Veh1_D_1 Veh2         Veh2_D_1 Veh3 Veh3_D_1
## 1                              0   No                0   No        0
## 2 Turkey Sandwich in boxed lunch  Yes Football players   No        0
## 3                              0   No                0   No        0
## 4                              0   No                0   No        0
## 5                              0   No                0   No        0
## 6                              0   No                0   No        0
##      PCRSect OBYear Hemisphere season MeanI1 MedianI1 Range_S_I1
## 1     Capsid   1999   Northern   Fall      0        0          0
## 2 Polymerase   1998   Northern   Fall      0       37          0
## 3       Both   2006   Northern   Fall      0        0          0
## 4       Both   2006   Northern   Fall      0        0          0
## 5       Both   2006   Northern   Fall      0        0          0
## 6       Both   2006   Northern   Fall      0        0          0
##   Range_L_I1 MeanD1 MedianD1 Range_S_D1 Range_L_D1 MeanA1 MedianA1
## 1          0      0        0          0          0     NA       NA
## 2          0      0       36          0          0     NA       NA
## 3          0      0        0          0          0     NA       NA
## 4          0      0        0          0          0     NA       NA
## 5          0      0        0          0          0     NA       NA
## 6          0      0        0          0          0     NA       NA
##   Range_Y_A1 Range_O_A1     Action1 Action2_1 Secondary MeanI2 MedianI2
## 1       0.75          2 Unspecified         0        No      0        0
## 2          0          0 Unspecified         0       Yes      0        0
## 3          0          0 Unspecified         0        No      0        0
## 4          0          0 Unspecified         0        No      0        0
## 5          0          0 Unspecified         0        No      0        0
## 6          0          0 Unspecified         0        No      0        0
##   Range_S_I2 Range_L_I2 MeanD2 MedianD2 Range_S_D2 Range_L_D2 Mea.2
## 1          0          0      0        0          0          0     0
## 2          0          0      0        0          0          0     0
## 3          0          0      0        0          0          0     0
## 4          0          0      0        0          0          0     0
## 5          0          0      0        0          0          0     0
## 6          0          0      0        0          0          0     0
##   Media.2 Range_Y_A2 Range_O_A2
## 1       0          0          0
## 2       0          0          0
## 3       0          0          0
## 4       0          0          0
## 5       0          0          0
## 6       0          0          0
##                                                                                                                  Comments_1
## 1                   Outbreak took place during a study on gasteroenteritus in a day care center. Same paper as outbreak # 2
## 2 Secondary cases include both persons from NC and FL, some secondary cases were included in # at risk of primary infection
## 3                                                                                         VWA outbreak no. 68592 in Table 1
## 4                                                                                         VWA outbreak no. 69113 in Table 1
## 5                                                                                         VWA outbreak no. 69479 in Table 1
## 6                                                                                         VWA outbreak no. 69490 in Table 1
##         Path1 Path2_1 Country    Category  State
## 1          No       0   Japan     Daycare      0
## 2          No       0     USA Foodservice NC, FL
## 3 Unspecified       0   Other Foodservice      0
## 4 Unspecified       0   Other Foodservice      0
## 5 Unspecified       0   Other Foodservice      0
## 6 Unspecified       0   Other Foodservice      0
##                    Setting_1 StartMonth EndMonth GGA CA                SA
## 1             Daycare Center         11       12   2  4         Lordsdale
## 2 Boxed lunch, football game          9        9   1  0 Thistle Hall 1/91
## 3                     buffet          9        0   2  4       GII.4 2006a
## 4                 restaurant         10        0   0  0                 0
## 5                     buffet         11        0   2  4       GII.4 2006b
## 6        take-out restaurant         11        0   0  0                 0
##   new_GGA new_CA new_SA SA_resolved_from GGB CB SB new_GGB new_CB new_SB
## 1       0      0      0                    0  0  0       0      0      0
## 2       0      0      0                    0  0  0       0      0      0
## 3       0      0      0                    0  0  0       0      0      0
## 4       0      0      0                    0  0  0       0      0      0
## 5       0      0      0                    0  0  0       0      0      0
## 6       0      0      0                    0  0  0       0      0      0
##   SB_resolved_from GGC CC SC new_ggc new_cc new_sc SC_resolved_from GGD CD
## 1                    0  0  0       0      0      0                    0  0
## 2                    0  0  0       0      0      0                    0  0
## 3                    0  0  0       0      0      0                    0  0
## 4                    0  0  0       0      0      0                    0  0
## 5                    0  0  0       0      0      0                    0  0
## 6                    0  0  0       0      0      0                    0  0
##   SD new_ggd new_cd new_sd SD_resolved_from StrainOther strainother_rc gge
## 1  0       0      0      0               NA           0              0   0
## 2  0       0      0      0               NA           0              0   0
## 3  0       0      0      0               NA           0              0   0
## 4  0       0      0      0               NA           0              0   0
## 5  0       0      0      0               NA           0              0   0
## 6  0       0      0      0               NA           0              0   0
##   ce se SE_resolved_from ggf cf sf ggg cg sg ggh ch sh ggi ci si ggj cj sj
## 1  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
## 2  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
## 3  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
## 4  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
## 5  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
## 6  0  0                    0  0  0   0  0  0   0  0  0   0  0  0   0  0  0
##          Country2    Veh1_D_2 Veh2_D_2 Veh3_D_2 Action2_2
## 1               0           0        0        0         0
## 2               0 Boxed Lunch        0        0         0
## 3 The Netherlands           0        0        0         0
## 4 The Netherlands           0        0        0         0
## 5 The Netherlands           0        0        0         0
## 6 The Netherlands           0        0        0         0
##                      Comments_2 Path2_2           Setting_2      category1
## 1                  Limited data       0                   0 School/Daycare
## 2                             0       0                   0    Foodservice
## 3 Outbreak 19 of 26 Boxman 2009       0              Buffet    Foodservice
## 4 Outbreak 20 of 26 Boxman 2009       0          Restaurant    Foodservice
## 5 Outbreak 21 of 26 Boxman 2009       0              Buffet    Foodservice
## 6 Outbreak 22 of 26 Boxman 2009       0 take out restaurant    Foodservice
##   strainothergg2c4 gg2c4 Vomit IncInd SymInd PooledLat PooledSym PooledAge
## 1                0   Yes     1      0      0         0         0         0
## 2                0           1      0      0        37        36         0
## 3                0   Yes     1      0      0         0         0         0
## 4                0           1      0      0         0         0         0
## 5                0   Yes     1      0      0         0         0         0
## 6                0           1      0      0         0         0         0
##   IndividualLatent IndividualSymptomatic
## 1               NA                      
## 2               NA                      
## 3               NA                      
## 4               NA                      
## 5               NA                      
## 6               NA</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="kw">str</span>(raw_data)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    1022 obs. of  139 variables:
##  $ id                   : int  2 17 39 40 41 42 43 44 67 74 ...
##  $ Author               : Factor w/ 217 levels &quot; kagawa-Okamoto&quot;,..: 3 9 16 16 16 16 16 16 25 28 ...
##  $ Pub_Year             : int  2005 2000 2009 2009 2009 2009 2009 2009 2009 1994 ...
##  $ pubmedid             : int  15841336 11071673 19205471 19205471 19205471 19205471 19205471 19205471 19440360 8202078 ...
##  $ EpiCurve             : Factor w/ 3 levels &quot;&quot;,&quot;N&quot;,&quot;Y&quot;: 3 3 2 2 2 2 2 2 2 2 ...
##  $ TDComment            : Factor w/ 11 levels &quot;&quot;,&quot;Just norovirus outbreak&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ AHComment            : Factor w/ 3 levels &quot;&quot;,&quot;confirmed&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Trans1               : Factor w/ 6 levels &quot;Environmental&quot;,..: 5 2 2 2 2 2 2 2 5 2 ...
##  $ Trans1_O             : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Trans2               : Factor w/ 6 levels &quot;  (not applicable)&quot;,..: 1 5 1 1 1 1 1 1 1 1 ...
##  $ Trans2_O             : Factor w/ 2 levels &quot;0&quot;,&quot;Direct contact with diarrhea or vomitus&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Trans3               : Factor w/ 6 levels &quot;  (not applicable)&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Trans3_O             : Factor w/ 2 levels &quot;0&quot;,&quot;smoking&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Risk1                : num  0 108 130 4 25 ...
##  $ Risk2                : num  NA NA NA NA NA NA NA NA NA NA ...
##  $ RiskAll              : num  0 108 130 4 25 ...
##  $ Cases1               : int  15 43 27 4 15 6 40 10 116 45 ...
##  $ Cases2               : int  NA 22 NA NA NA NA NA NA NA NA ...
##  $ CasesAll             : int  15 65 27 4 15 6 40 10 116 45 ...
##  $ Rate1                : num  NA 39.8 20.8 100 60 ...
##  $ Rate2                : num  NA NA NA NA NA NA NA NA NA NA ...
##  $ RateAll              : num  0 39.8 20.8 100 60 ...
##  $ Hospitalizations     : int  0 0 0 0 0 0 0 0 5 10 ...
##  $ Deaths               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Vehicle_1            : Factor w/ 126 levels &quot;0&quot;,&quot;Aerosolized vomit&quot;,..: 1 6 1 1 1 1 1 1 1 73 ...
##  $ Veh1                 : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 3 4 3 3 3 3 3 3 3 4 ...
##  $ Veh1_D_1             : Factor w/ 163 levels &quot;0&quot;,&quot;aerosilized&quot;,..: 1 159 1 1 1 1 1 1 1 107 ...
##  $ Veh2                 : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 1 2 1 1 1 1 1 1 1 1 ...
##  $ Veh2_D_1             : Factor w/ 63 levels &quot;&quot;,&quot;0&quot;,&quot;accommodation environment on day of arrival&quot;,..: 2 21 2 2 2 2 2 2 2 2 ...
##  $ Veh3                 : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Veh3_D_1             : Factor w/ 19 levels &quot;&quot;,&quot;0&quot;,&quot;Brandy S ps&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ PCRSect              : Factor w/ 4 levels &quot;Both&quot;,&quot;Capsid&quot;,..: 2 3 1 1 1 1 1 1 1 4 ...
##  $ OBYear               : Factor w/ 23 levels &quot;0&quot;,&quot;1983&quot;,&quot;1990&quot;,..: 11 10 19 19 19 19 19 19 17 5 ...
##  $ Hemisphere           : Factor w/ 3 levels &quot;Northern&quot;,&quot;Southern&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ season               : Factor w/ 5 levels &quot;&quot;,&quot;Fall&quot;,&quot;Spring&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ MeanI1               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianI1             : int  0 37 0 0 0 0 0 0 0 31 ...
##  $ Range_S_I1           : num  0 0 0 0 0 0 0 0 0 2 ...
##  $ Range_L_I1           : num  0 0 0 0 0 0 0 0 0 69 ...
##  $ MeanD1               : num  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianD1             : num  0 36 0 0 0 0 0 0 0 48 ...
##  $ Range_S_D1           : num  0 0 0 0 0 0 0 0 0 10 ...
##  $ Range_L_D1           : int  0 0 0 0 0 0 0 0 0 168 ...
##  $ MeanA1               : num  NA NA NA NA NA NA NA NA NA NA ...
##  $ MedianA1             : num  NA NA NA NA NA NA NA NA NA NA ...
##  $ Range_Y_A1           : Factor w/ 49 levels &quot;&lt;1&quot;,&quot;0&quot;,&quot;0.167&quot;,..: 8 2 2 2 2 2 2 2 2 2 ...
##  $ Range_O_A1           : num  2 0 0 0 0 0 0 0 0 0 ...
##  $ Action1              : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 3 3 3 3 3 3 3 3 3 4 ...
##  $ Action2_1            : Factor w/ 186 levels &quot;&quot;,&quot;\tContami ted batch withdrawn from market; Finnish authorities issued alert through Rapid Alert System for Food&quot;| __truncated__,..: 5 5 5 5 5 5 5 5 5 108 ...
##  $ Secondary            : Factor w/ 3 levels &quot;&quot;,&quot;No&quot;,&quot;Yes&quot;: 2 3 2 2 2 2 2 2 2 2 ...
##  $ MeanI2               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianI2             : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_S_I2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_L_I2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MeanD2               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianD2             : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_S_D2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_L_D2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Mea.2                : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Media.2              : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_Y_A2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Range_O_A2           : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ Comments_1           : Factor w/ 989 levels &quot;&quot;,&quot; turally recombi nt GII NV outbreak occurred in infant home in Sapporo, Japan where only residents who were up &quot;| __truncated__,..: 918 938 978 979 980 981 984 982 350 853 ...
##  $ Path1                : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 1 1 3 3 3 3 3 3 1 3 ...
##  $ Path2_1              : Factor w/ 59 levels &quot;&quot;,&quot;0&quot;,&quot;adenovirus&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Country              : Factor w/ 22 levels &quot;Australia&quot;,&quot;Austria&quot;,..: 12 22 17 17 17 17 17 17 17 22 ...
##  $ Category             : Factor w/ 12 levels &quot;Daycare&quot;,&quot;Foodservice&quot;,..: 1 2 2 2 2 2 2 2 6 12 ...
##  $ State                : Factor w/ 36 levels &quot;0&quot;,&quot;1&quot;,&quot;14 states: CA, UT, KS, WI, IL, IN, OH, GA, FL, NC, VA, WV, NY, PA,&quot;,..: 1 21 1 1 1 1 1 1 1 9 ...
##  $ Setting_1            : Factor w/ 388 levels &quot;\tPsychiatric Care Center adjoined&quot;,..: 121 32 39 310 39 352 39 310 201 3 ...
##  $ StartMonth           : int  11 9 9 10 11 11 11 11 11 11 ...
##  $ EndMonth             : int  12 9 0 0 0 0 0 0 11 11 ...
##  $ GGA                  : int  2 1 2 0 2 0 0 0 2 0 ...
##  $ CA                   : int  4 0 4 0 4 0 0 0 4 0 ...
##  $ SA                   : Factor w/ 123 levels &quot;0&quot;,&quot;100% identity w/AB05308, 96% identity w/MOH&quot;,..: 62 115 34 1 35 1 1 1 122 1 ...
##  $ new_GGA              : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_CA               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_SA               : Factor w/ 27 levels &quot;0&quot;,&quot;ARG320-USA&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ SA_resolved_from     : Factor w/ 9 levels &quot;&quot;,&quot;abstraction&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ GGB                  : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ CB                   : Factor w/ 18 levels &quot;0&quot;,&quot;1&quot;,&quot;10&quot;,&quot;12&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ SB                   : Factor w/ 38 levels &quot;0&quot;,&quot;93.7% homology w/Saitama&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ new_GGB              : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_CB               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_SB               : Factor w/ 14 levels &quot;0&quot;,&quot;Birmingham&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ SB_resolved_from     : Factor w/ 6 levels &quot;&quot;,&quot;abstraction&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ GGC                  : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ CC                   : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ SC                   : Factor w/ 25 levels &quot;0&quot;,&quot;95% homology with Sundsvall&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ new_ggc              : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_cc               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_sc               : Factor w/ 5 levels &quot;0&quot;,&quot;Fayetteville&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ SC_resolved_from     : Factor w/ 4 levels &quot;&quot;,&quot;abstraction&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ GGD                  : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ CD                   : Factor w/ 13 levels &quot;0&quot;,&quot;1&quot;,&quot;12&quot;,&quot;14&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ SD                   : Factor w/ 8 levels &quot;0&quot;,&quot;Amsterdam&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ new_ggd              : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_cd               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ new_sd               : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ SD_resolved_from     : logi  NA NA NA NA NA NA ...
##   [list output truncated]</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw">summary</span>(raw_data)</a></code></pre></div>
<pre><code>##        id              Author       Pub_Year       pubmedid       
##  Min.   :   1.0   Iritani : 83   Min.   :1993   Min.   : 7514927  
##  1st Qu.: 256.2   Kageyama: 63   1st Qu.:2003   1st Qu.:12938042  
##  Median : 511.5   Lysen   : 60   Median :2005   Median :16231127  
##  Mean   : 511.6   Ozawa   : 55   Mean   :2005   Mean   :15890029  
##  3rd Qu.: 766.8   Hamano  : 43   3rd Qu.:2008   3rd Qu.:18495859  
##  Max.   :1024.0   Fukuda  : 39   Max.   :2011   Max.   :21537761  
##                   (Other) :679                  NA&#39;s   :3         
##  EpiCurve                                                     TDComment  
##   :  7                                                             :994  
##  N:784    Not sure                                                 : 12  
##  Y:231    Not found in folder of articles                          :  6  
##           Not found on Pubmed                                      :  3  
##           Just norovirus outbreak                                  :  1  
##           No vomiting in some outbreaks but those are not specified:  1  
##           (Other)                                                  :  5  
##                AHComment                 Trans1       Trans1_O
##                     :1012   Environmental   : 12   Min.   :0  
##  confirmed          :   8   Foodborne       :391   1st Qu.:0  
##  probably not usable:   2   Person to Person:166   Median :0  
##                             Unknown         : 63   Mean   :0  
##                             Unspecified     :318   3rd Qu.:0  
##                             Waterborne      : 72   Max.   :0  
##                                                               
##                 Trans2                                       Trans2_O   
##    (not applicable):928   0                                      :1021  
##  Environmental     : 42   Direct contact with diarrhea or vomitus:   1  
##  Foodborne         :  8                                                 
##  Other             :  1                                                 
##  Person to Person  : 42                                                 
##  Waterborne        :  1                                                 
##                                                                         
##                 Trans3        Trans3_O        Risk1         
##    (not applicable):1007   0      :1021   Min.   :    0.00  
##  Environmental     :   7   smoking:   1   1st Qu.:    0.00  
##  Foodborne         :   4                  Median :   10.00  
##  Other             :   1                  Mean   :  374.57  
##  Person to Person  :   2                  3rd Qu.:   95.75  
##  Waterborne        :   1                  Max.   :35000.00  
##                                                             
##      Risk2           RiskAll            Cases1            Cases2       
##  Min.   : 66.00   Min.   :    0.0   Min.   :    1.0   Min.   :   1.00  
##  1st Qu.: 72.25   1st Qu.:    0.0   1st Qu.:    9.0   1st Qu.:   4.25  
##  Median : 82.00   Median :   22.5   Median :   26.0   Median :  10.50  
##  Mean   :135.06   Mean   :  425.6   Mean   :  125.2   Mean   :  53.02  
##  3rd Qu.:210.59   3rd Qu.:  119.5   3rd Qu.:   64.5   3rd Qu.:  26.75  
##  Max.   :258.00   Max.   :35000.0   Max.   :32150.0   Max.   :1113.00  
##  NA&#39;s   :1014     NA&#39;s   :120       NA&#39;s   :7         NA&#39;s   :976      
##     CasesAll           Rate1              Rate2          RateAll      
##  Min.   :    1.0   Min.   :  0.4074   Min.   : 6.00   Min.   :  0.00  
##  1st Qu.:    9.0   1st Qu.: 17.2500   1st Qu.:10.70   1st Qu.:  0.00  
##  Median :   26.0   Median : 37.5000   Median :17.50   Median : 15.28  
##  Mean   :  127.6   Mean   : 40.9253   Mean   :16.89   Mean   : 26.65  
##  3rd Qu.:   66.0   3rd Qu.: 59.1875   3rd Qu.:21.90   3rd Qu.: 48.91  
##  Max.   :32150.0   Max.   :100.0000   Max.   :29.69   Max.   :105.00  
##  NA&#39;s   :7         NA&#39;s   :431        NA&#39;s   :1012    NA&#39;s   :108     
##  Hospitalizations       Deaths                   Vehicle_1  
##  Min.   :  0.0000   Min.   :0.00000   0               :694  
##  1st Qu.:  0.0000   1st Qu.:0.00000   Oysters         : 46  
##  Median :  0.0000   Median :0.00000   oysters         : 42  
##  Mean   :  0.8856   Mean   :0.05005   Shellfish       : 21  
##  3rd Qu.:  0.0000   3rd Qu.:0.00000   lettuce         : 11  
##  Max.   :197.0000   Max.   :9.00000   Infected Persons: 10  
##  NA&#39;s   :43         NA&#39;s   :43        (Other)         :198  
##           Veh1                 Veh1_D_1    Veh2    
##  No         :  1   0               :575   No :933  
##  Unknown    : 82   Infected Persons: 60   Yes: 89  
##  Unspecified:493   Oysters         : 47            
##  Yes        :446   oysters         : 42            
##                    Shellfish       : 20            
##                    Food            : 14            
##                    (Other)         :264            
##                    Veh2_D_1    Veh3                      Veh3_D_1  
##  0                     :814   No :1004   0                   :895  
##                        :120   Yes:  18                       :109  
##  Infected Persons      : 12              Contami ted Surfaces:  2  
##  Contami ted Surfaces  :  8              Brandy S ps         :  1  
##  salad                 :  4              Chicken             :  1  
##  Vomit or Stool contact:  3              Coffee White        :  1  
##  (Other)               : 61              (Other)             : 13  
##         PCRSect        OBYear          Hemisphere     season   
##  Both       :441   2006   :142   Northern   :939         : 67  
##  Capsid     :138   2002   :134   Southern   : 81   Fall  :164  
##  Polymerase :175   2004   :116   Unspecified:  2   Spring:222  
##  Unspecified:268   2003   : 97                     Summer:126  
##                    1999   : 80                     Winter:443  
##                    1998   : 76                                 
##                    (Other):377                                 
##      MeanI1           MedianI1        Range_S_I1       Range_L_I1     
##  Min.   : 0.0000   Min.   : 0.000   Min.   : 0.000   Min.   :  0.000  
##  1st Qu.: 0.0000   1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.:  0.000  
##  Median : 0.0000   Median : 0.000   Median : 0.000   Median :  0.000  
##  Mean   : 0.6644   Mean   : 1.588   Mean   : 1.066   Mean   :  3.976  
##  3rd Qu.: 0.0000   3rd Qu.: 0.000   3rd Qu.: 0.000   3rd Qu.:  0.000  
##  Max.   :48.0000   Max.   :65.000   Max.   :36.000   Max.   :168.000  
##                                                                       
##      MeanD1           MedianD1         Range_S_D1       Range_L_D1    
##  Min.   :  0.000   Min.   :  0.000   Min.   : 0.000   Min.   :   0.0  
##  1st Qu.:  0.000   1st Qu.:  0.000   1st Qu.: 0.000   1st Qu.:   0.0  
##  Median :  0.000   Median :  0.000   Median : 0.000   Median :   0.0  
##  Mean   :  1.579   Mean   :  2.552   Mean   : 1.486   Mean   :  12.7  
##  3rd Qu.:  0.000   3rd Qu.:  0.000   3rd Qu.: 0.000   3rd Qu.:   0.0  
##  Max.   :273.600   Max.   :235.200   Max.   :72.000   Max.   :1248.0  
##                                                                       
##      MeanA1           MedianA1        Range_Y_A1    Range_O_A1     
##  Min.   : 0.0833   Min.   : 0.747   0      :891   Min.   :  0.000  
##  1st Qu.:13.0000   1st Qu.: 9.000   1      : 10   1st Qu.:  0.000  
##  Median :39.0000   Median :29.000   18     :  9   Median :  0.000  
##  Mean   :38.2923   Mean   :29.485   23     :  7   Mean   :  7.077  
##  3rd Qu.:56.0000   3rd Qu.:45.000   16     :  6   3rd Qu.:  0.000  
##  Max.   :89.3300   Max.   :85.000   20     :  6   Max.   :103.000  
##  NA&#39;s   :961       NA&#39;s   :973      (Other): 93                    
##         Action1   
##  No         :  3  
##  Unknown    :  2  
##  Unspecified:803  
##  Yes        :214  
##                   
##                   
##                   
##                                                                                                                                                                                                                                         Action2_1  
##  0                                                                                                                                                                                                                                           :807  
##  recalled lettuce from wholesaler                                                                                                                                                                                                            : 11  
##  Infection control practices essation of occupatio l therapy; cohort program;  visiting restrictions; symptomatic staff sent home                                                                                                            :  4  
##  Restaurant was cleaned, closed, then professio lly cleaned                                                                                                                                                                                  :  4  
##  Camp was clean, closed and the fi lly professio lly cleaned                                                                                                                                                                                 :  3  
##  exclude members of affected groups from water sports, activities with shared implements; provided dedicated latrines, washing facilities, drinking water; excluded from food handling, preparation; isolated until asymptomatic for 48 hours:  3  
##  (Other)                                                                                                                                                                                                                                     :190  
##  Secondary     MeanI2           MedianI2          Range_S_I2     
##     :  1   Min.   : 0.0000   Min.   : 0.00000   Min.   : 0.0000  
##  No :972   1st Qu.: 0.0000   1st Qu.: 0.00000   1st Qu.: 0.0000  
##  Yes: 49   Median : 0.0000   Median : 0.00000   Median : 0.0000  
##            Mean   : 0.1203   Mean   : 0.04697   Mean   : 0.2172  
##            3rd Qu.: 0.0000   3rd Qu.: 0.00000   3rd Qu.: 0.0000  
##            Max.   :81.0000   Max.   :48.00000   Max.   :78.0000  
##                                                                  
##    Range_L_I2           MeanD2     MedianD2   Range_S_D2      
##  Min.   :  0.0000   Min.   :0   Min.   :0   Min.   : 0.00000  
##  1st Qu.:  0.0000   1st Qu.:0   1st Qu.:0   1st Qu.: 0.00000  
##  Median :  0.0000   Median :0   Median :0   Median : 0.00000  
##  Mean   :  0.4511   Mean   :0   Mean   :0   Mean   : 0.02348  
##  3rd Qu.:  0.0000   3rd Qu.:0   3rd Qu.:0   3rd Qu.: 0.00000  
##  Max.   :168.0000   Max.   :0   Max.   :0   Max.   :24.00000  
##                                                               
##    Range_L_D2           Mea.2             Media.2        
##  Min.   : 0.00000   Min.   : 0.00000   Min.   : 0.00000  
##  1st Qu.: 0.00000   1st Qu.: 0.00000   1st Qu.: 0.00000  
##  Median : 0.00000   Median : 0.00000   Median : 0.00000  
##  Mean   : 0.09393   Mean   : 0.08904   Mean   : 0.04795  
##  3rd Qu.: 0.00000   3rd Qu.: 0.00000   3rd Qu.: 0.00000  
##  Max.   :72.00000   Max.   :55.00000   Max.   :49.00000  
##                                                          
##    Range_Y_A2         Range_O_A2     
##  Min.   : 0.00000   Min.   : 0.0000  
##  1st Qu.: 0.00000   1st Qu.: 0.0000  
##  Median : 0.00000   Median : 0.0000  
##  Mean   : 0.05773   Mean   : 0.2211  
##  3rd Qu.: 0.00000   3rd Qu.: 0.0000  
##  Max.   :46.00000   Max.   :80.0000  
##                                      
##                                         Comments_1          Path1    
##                                              : 14   No         :223  
##  0                                           : 10   Unknown    :  3  
##  2 of 3, paper about genetic a lysis, 1 death:  2   Unspecified:716  
##  3 of 3, limited information available       :  2   Yes        : 80  
##  Outbreak 10 of 47, limited information      :  2                    
##  outbreak A in Table 1                       :  2                    
##  (Other)                                     :990                    
##                  Path2_1           Country            Category  
##  0                   :833   Other      :447   Foodservice :309  
##                      :109   Japan      :371   Leisure     :153  
##  coliforms           :  8   USA        :108   Nursing Home:138  
##  Astrovirus          :  4   Australia  : 21   Hospital    :126  
##  E.coli              :  4   Multiple   : 17   Other       : 99  
##  Campylobacter jejuni:  3   Unspecified: 15   School      : 68  
##  (Other)             : 61   (Other)    : 43   (Other)     :129  
##      State                   Setting_1     StartMonth        EndMonth     
##  0      :909   0                  :110   Min.   : 0.000   Min.   : 0.000  
##  NC     : 19   Restaurant         :104   1st Qu.: 2.000   1st Qu.: 0.000  
##  NY     : 10   restaurant         : 70   Median : 4.000   Median : 0.000  
##  AK     :  9   Hospital           : 39   Mean   : 5.532   Mean   : 2.414  
##  FL     :  7   Hotel              : 25   3rd Qu.:10.000   3rd Qu.: 3.000  
##  MI     :  6   Nursing Care Center: 25   Max.   :12.000   Max.   :12.000  
##  (Other): 62   (Other)            :649                                    
##       GGA              CA                        SA         new_GGA      
##  Min.   :0.000   Min.   : 0.000   0               :693   Min.   :0.0000  
##  1st Qu.:1.000   1st Qu.: 0.000   Lordsdale       : 45   1st Qu.:0.0000  
##  Median :2.000   Median : 2.000   Farmington Hills: 18   Median :0.0000  
##  Mean   :1.501   Mean   : 2.551   GII.4 2006b     : 16   Mean   :0.1693  
##  3rd Qu.:2.000   3rd Qu.: 4.000   GII.4 2006a     : 15   3rd Qu.:0.0000  
##  Max.   :2.000   Max.   :17.000   Desert Shield   : 14   Max.   :2.0000  
##                  NA&#39;s   :1        (Other)         :221                   
##      new_CA                 new_SA   
##  Min.   :0.0000   0            :926  
##  1st Qu.:0.0000   Toronto-CAN93: 15  
##  Median :0.0000   Grimsby-like : 11  
##  Mean   :0.2984   Hawaii-USA94 : 10  
##  3rd Qu.:0.0000   CBW94-AUS    :  7  
##  Max.   :9.0000   Chiba-JPN00  :  7  
##                   (Other)      : 46  
##                                                                         SA_resolved_from
##                                                                                 :926    
##  Zheng                                                                          : 69    
##  abstraction                                                                    : 17    
##  http://onlinelibrary.wiley.com/doi/10.1002/jmv.20403/pdf                       :  3    
##  origi l article                                                                :  3    
##  Diagnosis of Norovirus outbreaks by commercial ELISA or RT-PCR, de Bruin et al.:  1    
##  (Other)                                                                        :  3    
##       GGB               CB                SB         new_GGB      
##  Min.   :0.0000   0      :886   0          :976   Min.   :0.0000  
##  1st Qu.:0.0000   4      : 36   Lordsdale  :  3   1st Qu.:0.0000  
##  Median :0.0000   3      : 22   Toronto    :  3   Median :0.0000  
##  Mean   :0.3082   2      : 15   Birmingham :  2   Mean   :0.0274  
##  3rd Qu.:0.0000   6      : 15   GII.4 2006b:  2   3rd Qu.:0.0000  
##  Max.   :2.0000   1      :  9   GII.b      :  2   Max.   :2.0000  
##                   (Other): 39   (Other)    : 34                   
##      new_CB                   new_SB    
##  Min.   : 0.00000   0            :1005  
##  1st Qu.: 0.00000   Birmingham   :   2  
##  Median : 0.00000   Chiba-JPN00  :   2  
##  Mean   : 0.06947   Lsdale-GBR   :   2  
##  3rd Qu.: 0.00000   Toronto-CAN93:   2  
##  Max.   :14.00000   Bristol-GBR93:   1  
##                     (Other)      :   8  
##                                                                                                                                           SB_resolved_from
##                                                                                                                                                   :1005   
##  abstraction                                                                                                                                      :   2   
##  Evaluation of a broadly reactive nucleic acid sequence based amplification assay for the detection of noroviruses in faecal material, Moore et al:   2   
##  http://kcdc.labkm.net/vsd/database/gene_list.jsp?orgId=6&amp;cuTag=CL0005&amp;page=910&amp;orderCol=&amp;orderTag=&amp;searchIn=&amp;textIn=                             :   1   
##  origi l article/Zheng                                                                                                                            :   2   
##  Zheng                                                                                                                                            :  10   
##                                                                                                                                                           
##       GGC               CC        
##  Min.   :0.0000   Min.   : 0.000  
##  1st Qu.:0.0000   1st Qu.: 0.000  
##  Median :0.0000   Median : 0.000  
##  Mean   :0.1223   Mean   : 0.365  
##  3rd Qu.:0.0000   3rd Qu.: 0.000  
##  Max.   :2.0000   Max.   :17.000  
##                                   
##                                                                                                    SC     
##  0                                                                                                  :997  
##  H104-94-J                                                                                          :  2  
##  95% homology with Sundsvall                                                                        :  1  
##  99 and 98% deduced amino acid identity with FAY=98 and KAS/00 , respectively (both 92% nt identity):  1  
##  BCCDC04-007                                                                                        :  1  
##  Chiba                                                                                              :  1  
##  (Other)                                                                                            : 19  
##     new_ggc             new_cc                 new_sc    
##  Min.   :0.000000   Min.   : 0.0000   0           :1018  
##  1st Qu.:0.000000   1st Qu.: 0.0000   Fayetteville:   1  
##  Median :0.000000   Median : 0.0000   Leeds-GBR00 :   1  
##  Mean   :0.006849   Mean   : 0.0274   MOH99-HUN   :   1  
##  3rd Qu.:0.000000   3rd Qu.: 0.0000   Whiterose   :   1  
##  Max.   :2.000000   Max.   :14.0000                      
##                                                          
##                                                                                                                                           SC_resolved_from
##                                                                                                                                                   :1018   
##  abstraction                                                                                                                                      :   1   
##  Evaluation of a broadly reactive nucleic acid sequence based amplification assay for the detection of noroviruses in faecal material, Moore et al:   1   
##  Zheng                                                                                                                                            :   2   
##                                                                                                                                                           
##                                                                                                                                                           
##                                                                                                                                                           
##       GGD                CD                                  SD      
##  Min.   :0.00000   0      :990   0                            :1014  
##  1st Qu.:0.00000   3      :  6   GII.b                        :   2  
##  Median :0.00000   4      :  5   Amsterdam                    :   1  
##  Mean   :0.06067   8      :  5   BCCDC03-008                  :   1  
##  3rd Qu.:0.00000   6      :  4   Bristol                      :   1  
##  Max.   :2.00000   15     :  3   Harrow/2001/UK_Mexico/1989/MX:   1  
##                    (Other):  9   (Other)                      :   2  
##     new_ggd      new_cd      new_sd  SD_resolved_from
##  Min.   :0   Min.   :0   Min.   :0   Mode:logical    
##  1st Qu.:0   1st Qu.:0   1st Qu.:0   NA&#39;s:1022       
##  Median :0   Median :0   Median :0                   
##  Mean   :0   Mean   :0   Mean   :0                   
##  3rd Qu.:0   3rd Qu.:0   3rd Qu.:0                   
##  Max.   :0   Max.   :0   Max.   :0                   
##                                                      
##                                                                                                                                                        StrainOther 
##  0                                                                                                                                                           :980  
##  closest relative is SRSV/Apeldoorn/97/NET w/98% nucleotide identity over 219 bp of the polymerase region and 96% (over 283 bp) was also seen w/strain Arg320:  2  
##  GII/8                                                                                                                                                       :  2  
##  GIIb                                                                                                                                                        :  2  
##  1                                                                                                                                                           :  1  
##  100% similar to the calicivirus strains of S031/94/UK                                                                                                       :  1  
##  (Other)                                                                                                                                                     : 34  
##                                                                                               strainother_rc
##  0                                                                                                   :1013  
##  closest relative is SRSV/Apeldoorn/97/NET w/98% nucleotide identity over 219 bp of the polymerase re:   2  
##  100% similar to the calicivirus strains of S031/94/UK                                               :   1  
##  2 strains only referred to as strain A and strain B                                                 :   1  
##  6                                                                                                   :   1  
##  article discussed 2 separate strains found, but didn&#39;t identify either                              :   1  
##  (Other)                                                                                             :   3  
##          gge            ce                     se      
##  0         :994   Min.   : 0.0000   0           :1012  
##  1         :  7   1st Qu.: 0.0000   GIIb        :   2  
##  2         : 20   Median : 0.0000   BCCDC04-002 :   1  
##  Sindlesham:  1   Mean   : 0.1272   Fayetteville:   1  
##                   3rd Qu.: 0.0000   GII.4 2006a :   1  
##                   Max.   :17.0000   GII.4a 2004 :   1  
##                                     (Other)     :   4  
##                    SE_resolved_from      ggf                cf          
##                            :1016    Min.   :0.00000   Min.   : 0.00000  
##  abstraction of StrainOther:   6    1st Qu.:0.00000   1st Qu.: 0.00000  
##                                     Median :0.00000   Median : 0.00000  
##                                     Mean   :0.02055   Mean   : 0.08317  
##                                     3rd Qu.:0.00000   3rd Qu.: 0.00000  
##                                     Max.   :2.00000   Max.   :14.00000  
##                                                                         
##            sf            ggg               cg                     sg      
##  0          :1021   Min.   :0.0000   Min.   : 0.00000   0          :1021  
##  BCCDC03-013:   1   1st Qu.:0.0000   1st Qu.: 0.00000   BCCDC04-006:   1  
##                     Median :0.0000   Median : 0.00000                     
##                     Mean   :0.0137   Mean   : 0.05186                     
##                     3rd Qu.:0.0000   3rd Qu.: 0.00000                     
##                     Max.   :2.0000   Max.   :12.00000                     
##                                                                           
##       ggh                 ch                     sh      
##  Min.   :0.000000   Min.   : 0.00000   0          :1020  
##  1st Qu.:0.000000   1st Qu.: 0.00000   BCCDC03-032:   1  
##  Median :0.000000   Median : 0.00000   GIIb       :   1  
##  Mean   :0.006849   Mean   : 0.02251                     
##  3rd Qu.:0.000000   3rd Qu.: 0.00000                     
##  Max.   :2.000000   Max.   :14.00000                     
##                                                          
##       ggi                 ci                     si      
##  Min.   :0.000000   Min.   :0.000000   0          :1021  
##  1st Qu.:0.000000   1st Qu.:0.000000   BCCDC04-003:   1  
##  Median :0.000000   Median :0.000000                     
##  Mean   :0.002935   Mean   :0.003914                     
##  3rd Qu.:0.000000   3rd Qu.:0.000000                     
##  Max.   :2.000000   Max.   :3.000000                     
##                                                          
##       ggj                 cj                     sj      
##  Min.   :0.000000   Min.   :0.000000   0          :1021  
##  1st Qu.:0.000000   1st Qu.:0.000000   BCCDC04-008:   1  
##  Median :0.000000   Median :0.000000                     
##  Mean   :0.002935   Mean   :0.006849                     
##  3rd Qu.:0.000000   3rd Qu.:0.000000                     
##  Max.   :2.000000   Max.   :5.000000                     
##                                                          
##             Country2        Veh1_D_2  
##  0              :547   0        :711  
##  Sweden         : 71   Oyster   : 68  
##  United Kingdom : 51   Shellfish: 23  
##  The Netherlands: 49   oyster   : 16  
##  Australia      : 41   Oysters  : 12  
##  Finland        : 26   lettuce  : 11  
##  (Other)        :237   (Other)  :181  
##                                     Veh2_D_2            Veh3_D_2   
##  0                                      :1000   0           :1020  
##  Salad                                  :   3   house salad :   1  
##  areas contami ted with index case vomit:   1   udon noodles:   1  
##  assymptomatic foodhandler              :   1                      
##  buffet spring rolls                    :   1                      
##  coleslaw                               :   1                      
##  (Other)                                :  15                      
##                                                                                                                                           Action2_2  
##  0                                                                                                                                             :887  
##  groups with affected members were excluded from camp activities and limited to specific bathrooms and water fountains                         :  3  
##  professio lly cleaned restaurant                                                                                                              :  3  
##  1 week sanitation and sterilization of the ship after cruise 1, however outbreak continued on the next 5 cruises                              :  1  
##  admission to the hospital closed 3 times, 34/41 guidlines for NoV outbreak control covered,  control implemented via the hospital hygiene team:  1  
##  admissions suspended, handwashing, enviornmental cleaning, visitor restriction, linen cleaning, staff policy etc.                             :  1  
##  (Other)                                                                                                                                       :126  
##                                                                                                                                                     Comments_2 
##  0                                                                                                                                                       :318  
##                                                                                                                                                          : 22  
##  Hosp A ward 7C                                                                                                                                          :  2  
##  Hosp A ward 7E                                                                                                                                          :  2  
##  Outbreak 13 of 19 Iritani 2008                                                                                                                          :  2  
##  1 of 12 outbreaks reported in shinkawa article; cases/persons at risk and the subsequent attack rate includes the food handler data presented in table 1:  1  
##  (Other)                                                                                                                                                 :675  
##                   Path2_2          Setting_2            category1  
##  0                    :976   0          :350   Foodservice   :309  
##  Kobuvirus            :  4   Restaurant :154   Hospital/Nursi:225  
##  Adenovirus           :  2              : 22   Leisure       :153  
##  astrovirus           :  2   Hotel      : 18   Other         :120  
##  Clostridium difficile:  2   cruise ship: 14   School/Daycare: 84  
##  rotavirus            :  2   multiple   : 11   Unknown       : 74  
##  (Other)              : 34   (Other)    :453   (Other)       : 57  
##  strainothergg2c4   gg2c4         Vomit            IncInd        
##  Min.   :0.000000      :716   Min.   :0.0000   Min.   :0.000000  
##  1st Qu.:0.000000   Yes:306   1st Qu.:0.0000   1st Qu.:0.000000  
##  Median :0.000000             Median :0.0000   Median :0.000000  
##  Mean   :0.005871             Mean   :0.4907   Mean   :0.001957  
##  3rd Qu.:0.000000             3rd Qu.:1.0000   3rd Qu.:0.000000  
##  Max.   :1.000000             Max.   :1.0000   Max.   :1.000000  
##                               NA&#39;s   :1                          
##      SymInd           PooledLat        PooledSym        PooledAge     
##  Min.   :0.000000   Min.   : 0.000   Min.   : 0.000   Min.   : 0.000  
##  1st Qu.:0.000000   1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000  
##  Median :0.000000   Median : 0.000   Median : 0.000   Median : 0.000  
##  Mean   :0.005871   Mean   : 1.889   Mean   : 3.154   Mean   : 2.401  
##  3rd Qu.:0.000000   3rd Qu.: 0.000   3rd Qu.: 0.000   3rd Qu.: 0.000  
##  Max.   :1.000000   Max.   :67.500   Max.   :96.000   Max.   :85.000  
##                                                       NA&#39;s   :2       
##  IndividualLatent IndividualSymptomatic
##  Mode:logical      :1017               
##  NA&#39;s:1022        Y:   5               
##                                        
##                                        
##                                        
##                                        
## </code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="kw">class</span>(raw_data)</a></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
</div>
<div id="looking-at-the-outcome" class="section level1">
<h1>Looking at the outcome</h1>
<p>For this analysis, our main outcome of interest is if an outbreak was caused by the G2.4 strain of norovirus or not, and how other factors might be correlated with that strain.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co">#write code to take a look at the outcome variable (gg2c4)</span></a>
<a class="sourceLine" id="cb18-2" title="2"></a>
<a class="sourceLine" id="cb18-3" title="3"><span class="kw">glimpse</span>(raw_data<span class="op">$</span>gg2c4)</a></code></pre></div>
<pre><code>##  Factor w/ 2 levels &quot;&quot;,&quot;Yes&quot;: 2 1 2 1 2 1 1 1 2 1 ...</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="kw">summary</span>(raw_data<span class="op">$</span>gg2c4)</a></code></pre></div>
<pre><code>##     Yes 
## 716 306</code></pre>
<p>Overall, it looks ok, a decent amout in each category (i.e. no unbalanced data). However, we see an odd coding, there are only 2 types of entries, either “Yes” or "“, i.e. an empty space. The codebook tells us that this is how things were coded, if it was G2.4 it got a”Yes“, if it wasn’t it did get an empty space instead of a”No“. That’s somewhat strange and while the analysis should work, it is better to re-code the empty slots with”No".</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="co">#write code to change the empty values in gg2c4 to &quot;No&quot;</span></a>
<a class="sourceLine" id="cb22-2" title="2">gg2c4_revised &lt;-<span class="st"> </span>raw_data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb22-3" title="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">gg2c4=</span> <span class="kw">fct_recode</span>(gg2c4, <span class="dt">Yes =</span> <span class="st">&quot;Yes&quot;</span>, <span class="dt">No =</span> <span class="st">&quot;&quot;</span>))</a>
<a class="sourceLine" id="cb22-4" title="4"></a>
<a class="sourceLine" id="cb22-5" title="5"><span class="kw">glimpse</span>(gg2c4_revised<span class="op">$</span>gg2c4)</a></code></pre></div>
<pre><code>##  Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 2 1 2 1 2 1 1 1 2 1 ...</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1"><span class="kw">summary</span>(gg2c4_revised<span class="op">$</span>gg2c4)</a></code></pre></div>
<pre><code>##  No Yes 
## 716 306</code></pre>
</div>
<div id="selecting-predictors" class="section level1">
<h1>Selecting predictors</h1>
<p>We will pick similar variables as previously, with some adjustments based on what we learned before. Keep the following variables: <code>Action1, CasesAll, Country, Deaths, GG2C4, Hemisphere, Hospitalizations,  MeanD1, MeanI1, MedianD1, MedianI1, OBYear, Path1, RiskAll, Season, Setting, Trans1, Vomit.</code></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1">picked_variables &lt;-<span class="st"> </span>gg2c4_revised <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(Action1, CasesAll, Country, Deaths, gg2c4, Hemisphere, Hospitalizations,  MeanD1, MeanI1, MedianD1, MedianI1, OBYear, Path1, RiskAll, season, Setting_<span class="dv">1</span>, Trans1, Vomit)</a>
<a class="sourceLine" id="cb26-2" title="2"></a>
<a class="sourceLine" id="cb26-3" title="3"><span class="kw">dim</span>(picked_variables)</a></code></pre></div>
<pre><code>## [1] 1022   18</code></pre>
</div>
<div id="cleaning-predictors" class="section level1">
<h1>Cleaning predictors</h1>
<p>We’ll likely have to perform similar cleaning steps as before. A difference is that earlier we had to drop about half of the observations because no outcome was available. Here, we have outcome for every outbreak. Thus, there is more data to clean, which - as we will see - introduces a few issues we didn’t have before, since we dropped the troublesome observations early in the process.</p>
<p>Let’s first check for missing values.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="co"># write code that looks at missing values</span></a>
<a class="sourceLine" id="cb28-2" title="2"></a>
<a class="sourceLine" id="cb28-3" title="3"><span class="kw">vis_miss</span>(picked_variables, <span class="dt">sort_miss=</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/check-reduced-data-1.png" width="672" /></p>
<p>Looks like we have again some missing in <code>Hospitalization</code> and <code>Deaths</code>, and some more in <code>RiskAll</code>. Since the missing is not excessive, and to make our life easier, we’ll drop them for now. Note however the ‘blocks’ of missing values for RiskAll. Given that these outbreaks should be entered fairly randomly into the spreadsheet, it is strange to see the NA show up in such blocks. For a real data analysis, it would be worth looking closer and checking why there is a clustering like that.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1"><span class="co"># write code to remove any observations with NA</span></a>
<a class="sourceLine" id="cb29-2" title="2"></a>
<a class="sourceLine" id="cb29-3" title="3">missing_vars_gone &lt;-<span class="st"> </span><span class="kw">na.omit</span>(picked_variables)</a>
<a class="sourceLine" id="cb29-4" title="4"></a>
<a class="sourceLine" id="cb29-5" title="5"><span class="kw">vis_miss</span>(missing_vars_gone, <span class="dt">sort_miss=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/further-reduce-data-1.png" width="672" /></p>
<p>Let’s make sure everything has the right format (numeric/integer/factor). Adjust/recode variables as needed. You will likely find that as you convert <code>OBYear</code> to numeric, something doesn’t quite work. Take a look. Fix by removing the observation with the troublesome entry, then convert to numeric. Finally, remove the observations that have 0 as OByear - there are more than 1 now.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="co">#write code that cleans OBYear, convert it to numeric. Remove observations with OBYear = 0. </span></a>
<a class="sourceLine" id="cb30-2" title="2"><span class="co">#also convert any other variables as needed</span></a>
<a class="sourceLine" id="cb30-3" title="3"></a>
<a class="sourceLine" id="cb30-4" title="4"><span class="kw">vis_dat</span>(missing_vars_gone)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/adjust-1.png" width="672" /></p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1"><span class="kw">str</span>(missing_vars_gone)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    858 obs. of  18 variables:
##  $ Action1         : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 3 3 3 3 3 3 3 3 3 4 ...
##  $ CasesAll        : int  15 65 27 4 15 6 40 10 116 45 ...
##  $ Country         : Factor w/ 22 levels &quot;Australia&quot;,&quot;Austria&quot;,..: 12 22 17 17 17 17 17 17 17 22 ...
##  $ Deaths          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ gg2c4           : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 2 1 2 1 2 1 1 1 2 1 ...
##  $ Hemisphere      : Factor w/ 3 levels &quot;Northern&quot;,&quot;Southern&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Hospitalizations: int  0 0 0 0 0 0 0 0 5 10 ...
##  $ MeanD1          : num  0 0 0 0 0 0 0 0 0 0 ...
##  $ MeanI1          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianD1        : num  0 36 0 0 0 0 0 0 0 48 ...
##  $ MedianI1        : int  0 37 0 0 0 0 0 0 0 31 ...
##  $ OBYear          : Factor w/ 23 levels &quot;0&quot;,&quot;1983&quot;,&quot;1990&quot;,..: 11 10 19 19 19 19 19 19 17 5 ...
##  $ Path1           : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 1 1 3 3 3 3 3 3 1 3 ...
##  $ RiskAll         : num  0 108 130 4 25 ...
##  $ season          : Factor w/ 5 levels &quot;&quot;,&quot;Fall&quot;,&quot;Spring&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Setting_1       : Factor w/ 388 levels &quot;\tPsychiatric Care Center adjoined&quot;,..: 121 32 39 310 39 352 39 310 201 3 ...
##  $ Trans1          : Factor w/ 6 levels &quot;Environmental&quot;,..: 5 2 2 2 2 2 2 2 5 2 ...
##  $ Vomit           : int  1 1 1 1 1 1 1 1 1 1 ...
##  - attr(*, &quot;na.action&quot;)= &#39;omit&#39; Named int  20 35 37 38 86 93 98 99 100 101 ...
##   ..- attr(*, &quot;names&quot;)= chr  &quot;20&quot; &quot;35&quot; &quot;37&quot; &quot;38&quot; ...</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1">removed_0OBYear &lt;-<span class="st"> </span><span class="kw">subset</span>(missing_vars_gone, OBYear <span class="op">!=</span><span class="st"> &quot;0&quot;</span>)</a>
<a class="sourceLine" id="cb33-2" title="2">removed_0OBYear &lt;-<span class="st"> </span><span class="kw">droplevels</span>(removed_0OBYear)</a>
<a class="sourceLine" id="cb33-3" title="3"></a>
<a class="sourceLine" id="cb33-4" title="4"><span class="kw">str</span>(removed_0OBYear)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    852 obs. of  18 variables:
##  $ Action1         : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 3 3 3 3 3 3 3 3 3 4 ...
##  $ CasesAll        : int  15 65 27 4 15 6 40 10 116 45 ...
##  $ Country         : Factor w/ 5 levels &quot;Japan&quot;,&quot;Multiple&quot;,..: 1 5 3 3 3 3 3 3 3 5 ...
##  $ Deaths          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ gg2c4           : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 2 1 2 1 2 1 1 1 2 1 ...
##  $ Hemisphere      : Factor w/ 3 levels &quot;Northern&quot;,&quot;Southern&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Hospitalizations: int  0 0 0 0 0 0 0 0 5 10 ...
##  $ MeanD1          : num  0 0 0 0 0 0 0 0 0 0 ...
##  $ MeanI1          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianD1        : num  0 36 0 0 0 0 0 0 0 48 ...
##  $ MedianI1        : int  0 37 0 0 0 0 0 0 0 31 ...
##  $ OBYear          : Factor w/ 21 levels &quot;1983&quot;,&quot;1990&quot;,..: 10 9 17 17 17 17 17 17 15 4 ...
##  $ Path1           : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 1 1 3 3 3 3 3 3 1 3 ...
##  $ RiskAll         : num  0 108 130 4 25 ...
##  $ season          : Factor w/ 5 levels &quot;&quot;,&quot;Fall&quot;,&quot;Spring&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Setting_1       : Factor w/ 331 levels &quot;\tPsychiatric Care Center adjoined&quot;,..: 96 25 29 262 29 300 29 262 169 3 ...
##  $ Trans1          : Factor w/ 6 levels &quot;Environmental&quot;,..: 5 2 2 2 2 2 2 2 5 2 ...
##  $ Vomit           : int  1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1">year_as_numeric &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">as.character</span>(removed_0OBYear<span class="op">$</span>OBYear))</a>
<a class="sourceLine" id="cb35-2" title="2"></a>
<a class="sourceLine" id="cb35-3" title="3">removed_0OBYear<span class="op">$</span>OBYear &lt;-<span class="st"> </span>year_as_numeric</a>
<a class="sourceLine" id="cb35-4" title="4"><span class="kw">str</span>(removed_0OBYear)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    852 obs. of  18 variables:
##  $ Action1         : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 3 3 3 3 3 3 3 3 3 4 ...
##  $ CasesAll        : int  15 65 27 4 15 6 40 10 116 45 ...
##  $ Country         : Factor w/ 5 levels &quot;Japan&quot;,&quot;Multiple&quot;,..: 1 5 3 3 3 3 3 3 3 5 ...
##  $ Deaths          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ gg2c4           : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 2 1 2 1 2 1 1 1 2 1 ...
##  $ Hemisphere      : Factor w/ 3 levels &quot;Northern&quot;,&quot;Southern&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Hospitalizations: int  0 0 0 0 0 0 0 0 5 10 ...
##  $ MeanD1          : num  0 0 0 0 0 0 0 0 0 0 ...
##  $ MeanI1          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianD1        : num  0 36 0 0 0 0 0 0 0 48 ...
##  $ MedianI1        : int  0 37 0 0 0 0 0 0 0 31 ...
##  $ OBYear          : num  1999 1998 2006 2006 2006 ...
##  $ Path1           : Factor w/ 4 levels &quot;No&quot;,&quot;Unknown&quot;,..: 1 1 3 3 3 3 3 3 1 3 ...
##  $ RiskAll         : num  0 108 130 4 25 ...
##  $ season          : Factor w/ 5 levels &quot;&quot;,&quot;Fall&quot;,&quot;Spring&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Setting_1       : Factor w/ 331 levels &quot;\tPsychiatric Care Center adjoined&quot;,..: 96 25 29 262 29 300 29 262 169 3 ...
##  $ Trans1          : Factor w/ 6 levels &quot;Environmental&quot;,..: 5 2 2 2 2 2 2 2 5 2 ...
##  $ Vomit           : int  1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1"><span class="kw">vis_dat</span>(removed_0OBYear)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/adjust-2.png" width="672" /></p>
<p>Next, we remove the <code>Unspecified</code> entry in <code>Hemisphere</code> and recode <code>Action1</code> and <code>Path1</code> as described in the Data Analysis script, i.e. from <code>Unknown</code> to <code>Unspecified</code>. Also do the same grouping into just <code>Restaurant</code> and <code>Other</code> with the <code>Setting_1</code> variable. Again, remember that there are <code>restaurant</code> and <code>Restaurant</code> values, so you need to fix that too.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1"><span class="co"># write code that performs the actions described above</span></a>
<a class="sourceLine" id="cb38-2" title="2"><span class="co"># at the end, use the droplevels() command to remove empty factor levels</span></a>
<a class="sourceLine" id="cb38-3" title="3"></a>
<a class="sourceLine" id="cb38-4" title="4"><span class="co"># remove the unspecified entry and the level unspecified for hemisphere</span></a>
<a class="sourceLine" id="cb38-5" title="5"><span class="co">#glimpse(removed_0OBYear$Hemisphere)</span></a>
<a class="sourceLine" id="cb38-6" title="6">removeHemisphere &lt;-<span class="st"> </span>removed_0OBYear <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-7" title="7"><span class="st">  </span><span class="kw">subset</span>(Hemisphere <span class="op">!=</span><span class="st"> &quot;Unspecified&quot;</span>)</a>
<a class="sourceLine" id="cb38-8" title="8"></a>
<a class="sourceLine" id="cb38-9" title="9">removeHemisphere &lt;-<span class="st"> </span>removeHemisphere <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-10" title="10"><span class="st">  </span><span class="kw">droplevels</span>(removeHemisphere<span class="op">$</span>Hemisphere)</a>
<a class="sourceLine" id="cb38-11" title="11"></a>
<a class="sourceLine" id="cb38-12" title="12"><span class="co"># checking removeHemisphere let me see I only lost the one observation that was Unspecified for Hemisphere</span></a>
<a class="sourceLine" id="cb38-13" title="13"><span class="co"># look at responses and levels in Action variable</span></a>
<a class="sourceLine" id="cb38-14" title="14"><span class="co">#glimpse(removeHemisphere$Action1)</span></a>
<a class="sourceLine" id="cb38-15" title="15">combineAction1 &lt;-<span class="st"> </span>removeHemisphere <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-16" title="16"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Action1=</span> <span class="kw">fct_recode</span>(Action1, <span class="st">&quot;Unspecified&quot;</span> =<span class="st"> &quot;Unknown&quot;</span>))</a>
<a class="sourceLine" id="cb38-17" title="17"></a>
<a class="sourceLine" id="cb38-18" title="18"><span class="co">#glimpse(combineAction1$Path1)</span></a>
<a class="sourceLine" id="cb38-19" title="19">combinePath1 &lt;-<span class="st"> </span>combineAction1 <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-20" title="20"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Path1=</span> <span class="kw">fct_recode</span>(Path1, <span class="st">&quot;Unspecified&quot;</span> =<span class="st"> &quot;Unknown&quot;</span>))</a>
<a class="sourceLine" id="cb38-21" title="21"></a>
<a class="sourceLine" id="cb38-22" title="22"><span class="co">#here is what worked for me to label the restaurants</span></a>
<a class="sourceLine" id="cb38-23" title="23">restaurant_settings &lt;-<span class="st"> </span><span class="kw">subset</span>(combinePath1, Setting_<span class="dv">1</span> <span class="op">==</span><span class="st"> &quot;restaurant&quot;</span> <span class="op">|</span><span class="st"> </span>Setting_<span class="dv">1</span> <span class="op">==</span><span class="st"> &quot;Restaurant&quot;</span>)</a>
<a class="sourceLine" id="cb38-24" title="24">other_settings &lt;-<span class="st"> </span><span class="kw">subset</span>(combinePath1, Setting_<span class="dv">1</span> <span class="op">!=</span><span class="st"> &quot;restaurant&quot;</span> <span class="op">&amp;</span><span class="st"> </span>Setting_<span class="dv">1</span> <span class="op">!=</span><span class="st"> &quot;Restaurant&quot;</span>)</a>
<a class="sourceLine" id="cb38-25" title="25"></a>
<a class="sourceLine" id="cb38-26" title="26">restaurant_settings_edit &lt;-<span class="st"> </span>restaurant_settings <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-27" title="27"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Setting =</span> <span class="st">&quot;Restaurant&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-28" title="28"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(Setting_<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb38-29" title="29">  </a>
<a class="sourceLine" id="cb38-30" title="30">other_settings_edit &lt;-<span class="st"> </span>other_settings <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-31" title="31"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Setting =</span> <span class="st">&quot;Other&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-32" title="32"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(Setting_<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb38-33" title="33">  </a>
<a class="sourceLine" id="cb38-34" title="34"><span class="co">#here I incorporated gtools (library loaded at top) to use smartbind and paste both of my subsets together. It appears to have worked, as I am back with 579 rows again. Let&#39;s glance and see...</span></a>
<a class="sourceLine" id="cb38-35" title="35">further_cleaning &lt;-<span class="st"> </span><span class="kw">smartbind</span>(restaurant_settings_edit,other_settings_edit)</a>
<a class="sourceLine" id="cb38-36" title="36"><span class="co">#str(further_cleaning)</span></a>
<a class="sourceLine" id="cb38-37" title="37"></a>
<a class="sourceLine" id="cb38-38" title="38">further_cleaning<span class="op">$</span>Setting &lt;-<span class="st"> </span><span class="kw">as.factor</span>(further_cleaning<span class="op">$</span>Setting)</a>
<a class="sourceLine" id="cb38-39" title="39"></a>
<a class="sourceLine" id="cb38-40" title="40"><span class="kw">str</span>(further_cleaning)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    851 obs. of  18 variables:
##  $ Action1         : Factor w/ 3 levels &quot;No&quot;,&quot;Unspecified&quot;,..: 2 2 2 3 3 2 2 2 2 2 ...
##  $ CasesAll        : int  4 10 19 683 51 33 9 70 35 40 ...
##  $ Country         : Factor w/ 5 levels &quot;Japan&quot;,&quot;Multiple&quot;,..: 3 3 1 1 3 1 1 3 3 3 ...
##  $ Deaths          : int  0 0 0 0 0 0 0 0 0 0 ...
##  $ gg2c4           : Factor w/ 2 levels &quot;No&quot;,&quot;Yes&quot;: 1 1 1 1 1 1 2 2 1 2 ...
##  $ Hemisphere      : Factor w/ 2 levels &quot;Northern&quot;,&quot;Southern&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Hospitalizations: int  0 0 0 0 0 0 0 0 0 0 ...
##  $ MeanD1          : num  0 0 0 0 43 0 0 0 0 0 ...
##  $ MeanI1          : int  0 0 0 31 32 0 0 0 0 0 ...
##  $ MedianD1        : num  0 0 0 0 0 0 0 0 0 0 ...
##  $ MedianI1        : int  0 0 0 0 33 0 0 0 0 0 ...
##  $ OBYear          : num  2006 2006 2003 2003 2000 ...
##  $ Path1           : Factor w/ 3 levels &quot;No&quot;,&quot;Unspecified&quot;,..: 2 2 2 3 1 2 2 2 2 2 ...
##  $ RiskAll         : num  4 12 58 1492 72 ...
##  $ season          : Factor w/ 5 levels &quot;&quot;,&quot;Fall&quot;,&quot;Spring&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Trans1          : Factor w/ 6 levels &quot;Environmental&quot;,..: 2 2 2 2 2 2 2 2 2 2 ...
##  $ Vomit           : int  1 1 0 1 1 1 0 0 0 0 ...
##  $ Setting         : Factor w/ 2 levels &quot;Other&quot;,&quot;Restaurant&quot;: 2 2 2 2 2 2 2 2 2 2 ...</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1"><span class="kw">vis_dat</span>(further_cleaning)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/clean-more-predictors-1.png" width="672" /></p>
</div>
<div id="data-visualization" class="section level1">
<h1>Data visualization</h1>
<p>Next, let’s create a few plots showing the outcome and the predictors. For the continuous predictors, I suggest scatter/box/violinplots with the outcome on the x-axis.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" title="1"><span class="co">#write code that produces plots showing our outcome of interest on the x-axis and each numeric predictor on the y-axis.</span></a>
<a class="sourceLine" id="cb41-2" title="2"><span class="co">#you can use the facet_wrap functionality in ggplot for it, or do it some other way.</span></a>
<a class="sourceLine" id="cb41-3" title="3"></a>
<a class="sourceLine" id="cb41-4" title="4">p1 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, OBYear)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</a>
<a class="sourceLine" id="cb41-5" title="5"></a>
<a class="sourceLine" id="cb41-6" title="6">p2 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, RiskAll)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</a>
<a class="sourceLine" id="cb41-7" title="7"></a>
<a class="sourceLine" id="cb41-8" title="8">p3 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, MeanD1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_violin</span>()</a>
<a class="sourceLine" id="cb41-9" title="9"></a>
<a class="sourceLine" id="cb41-10" title="10">p4 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, MedianD1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_violin</span>()</a>
<a class="sourceLine" id="cb41-11" title="11"></a>
<a class="sourceLine" id="cb41-12" title="12"><span class="kw">grid.arrange</span>(p1, p2, p3, p4, <span class="dt">nrow=</span><span class="dv">2</span>, <span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/plots-1-1.png" width="672" /></p>
<p>Things look ok, apart from the skew in the predictors we discussed previously.</p>
<p>Next, let’s create plots for the categorical variabless. You can use for instance <code>geom_count</code> for it, or some other representation. If you prefer lots of tables, that’s ok too.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1"><span class="co">#write code that produces plots or tables showing our outcome of interest and each categorical predictor.</span></a>
<a class="sourceLine" id="cb42-2" title="2"></a>
<a class="sourceLine" id="cb42-3" title="3">p5 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Action1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-4" title="4">p6 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Path1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-5" title="5">p7 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Setting)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-6" title="6">p8 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Hemisphere)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-7" title="7">p9 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Country)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-8" title="8">p10 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, Trans1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-9" title="9">p11 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(further_cleaning, <span class="kw">aes</span>(gg2c4, season)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_count</span>()</a>
<a class="sourceLine" id="cb42-10" title="10"></a>
<a class="sourceLine" id="cb42-11" title="11"><span class="kw">grid.arrange</span>(p5, p6, p7, p8, p9, p10, p11, <span class="dt">ncol=</span><span class="dv">3</span>, <span class="dt">nrow=</span><span class="dv">3</span>, <span class="dt">widths=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="fl">1.15</span>))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/plots-2-1.png" width="672" /></p>
<p>You should see from plots or tables that some of the categories are small, e.g. for Action1, the “No” category is very small. Very few entries for a given factor create problems during cross-validation (since we can’t have a level show up in the holdout if it wasn’t part of the fitting set). So let’s look at those factor variables a bit closer and fix as needed.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="co">#write code that looks at tables/summaries of factors </span></a>
<a class="sourceLine" id="cb43-2" title="2"><span class="kw">summary</span>(further_cleaning)</a></code></pre></div>
<pre><code>##         Action1       CasesAll              Country        Deaths      
##  No         :  1   Min.   :    1.0   Japan      :316   Min.   :0.0000  
##  Unspecified:678   1st Qu.:    9.0   Multiple   : 16   1st Qu.:0.0000  
##  Yes        :172   Median :   25.0   Other      :409   Median :0.0000  
##                    Mean   :  127.8   Unspecified: 12   Mean   :0.0517  
##                    3rd Qu.:   65.0   USA        : 98   3rd Qu.:0.0000  
##                    Max.   :32150.0                     Max.   :9.0000  
##  gg2c4        Hemisphere  Hospitalizations      MeanD1      
##  No :592   Northern:798   Min.   : 0.0000   Min.   : 0.000  
##  Yes:259   Southern: 53   1st Qu.: 0.0000   1st Qu.: 0.000  
##                           Median : 0.0000   Median : 0.000  
##                           Mean   : 0.5511   Mean   : 1.309  
##                           3rd Qu.: 0.0000   3rd Qu.: 0.000  
##                           Max.   :99.0000   Max.   :96.000  
##      MeanI1           MedianD1         MedianI1         OBYear    
##  Min.   : 0.0000   Min.   : 0.000   Min.   : 0.00   Min.   :1983  
##  1st Qu.: 0.0000   1st Qu.: 0.000   1st Qu.: 0.00   1st Qu.:1999  
##  Median : 0.0000   Median : 0.000   Median : 0.00   Median :2002  
##  Mean   : 0.7156   Mean   : 2.309   Mean   : 1.72   Mean   :2002  
##  3rd Qu.: 0.0000   3rd Qu.: 0.000   3rd Qu.: 0.00   3rd Qu.:2005  
##  Max.   :48.0000   Max.   :72.000   Max.   :65.00   Max.   :2010  
##          Path1        RiskAll           season                 Trans1   
##  No         :196   Min.   :    0.0         : 56   Environmental   : 11  
##  Unspecified:615   1st Qu.:    0.0   Fall  :135   Foodborne       :338  
##  Yes        : 40   Median :   22.0   Spring:187   Person to Person:119  
##                    Mean   :  359.3   Summer:103   Unknown         : 61  
##                    3rd Qu.:  112.0   Winter:370   Unspecified     :261  
##                    Max.   :24000.0                Waterborne      : 61  
##      Vomit            Setting   
##  Min.   :0.00   Other     :692  
##  1st Qu.:0.00   Restaurant:159  
##  Median :0.00                   
##  Mean   :0.49                   
##  3rd Qu.:1.00                   
##  Max.   :1.00</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1"><span class="kw">table</span>(further_cleaning<span class="op">$</span>Action1)</a></code></pre></div>
<pre><code>## 
##          No Unspecified         Yes 
##           1         678         172</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1"><span class="kw">table</span>(further_cleaning<span class="op">$</span>Country)</a></code></pre></div>
<pre><code>## 
##       Japan    Multiple       Other Unspecified         USA 
##         316          16         409          12          98</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1"><span class="kw">table</span>(further_cleaning<span class="op">$</span>Trans1)</a></code></pre></div>
<pre><code>## 
##    Environmental        Foodborne Person to Person          Unknown 
##               11              338              119               61 
##      Unspecified       Waterborne 
##              261               61</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Action1))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-1.png" width="672" /></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Path1))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-2.png" width="672" /></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Setting))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-3.png" width="672" /></p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Hemisphere))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-4.png" width="672" /></p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Country))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-5.png" width="672" /></p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>Trans1))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-6.png" width="672" /></p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1"><span class="kw">barplot</span>(<span class="kw">table</span>(further_cleaning<span class="op">$</span>season))</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/more-factor-check-7.png" width="672" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" title="1"><span class="co">#further_cleaning</span></a></code></pre></div>
<p>You should see from your explorations above that there is only a single entry for <em>No</em> in <code>Action1</code>, and small entries for <em>Multiple</em> and <em>Unspecified</em> in <code>Country</code> and <em>Environmental</em> in <code>Trans1</code>. The single <em>No</em> should be fixed, the other somewhat small groupings might be ok. It depends on scientific rationale and method of analysis if you should modify/group those or not. We’ll do that.</p>
<p>Change things as follows: Remove the observation for which action is <em>No</em>, combine Countries into 3 groups Japan/USA/Other, and since I don’t even know what biologically the difference is between <em>Environmental</em> and <em>Waterborne</em> transmission (seems the same route to me based on my norovirus knowledge), we’ll move the Waterborne into the environmental. Finally, I noticed there is a blank category for season. That likely means it wasn’t stated in the paper. Let’s recode that as <em>Unknown</em>. Finally, re-order data such that the outcome is the first column and again remove empty factor levels. Then look at your resulting data frame.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" title="1"><span class="co">#write code that does the actions described above</span></a>
<a class="sourceLine" id="cb59-2" title="2"></a>
<a class="sourceLine" id="cb59-3" title="3"><span class="kw">summary</span>(further_cleaning<span class="op">$</span>Action1)</a></code></pre></div>
<pre><code>##          No Unspecified         Yes 
##           1         678         172</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" title="1"><span class="co">#Get rid of the No observation</span></a>
<a class="sourceLine" id="cb61-2" title="2">removed_Action1No &lt;-<span class="st"> </span>further_cleaning <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb61-3" title="3"><span class="st">  </span><span class="kw">subset</span>(Action1 <span class="op">!=</span><span class="st"> &quot;No&quot;</span>)</a>
<a class="sourceLine" id="cb61-4" title="4"></a>
<a class="sourceLine" id="cb61-5" title="5">removed_Action1No &lt;-<span class="st"> </span>removed_Action1No <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb61-6" title="6"><span class="st">  </span><span class="kw">droplevels</span>(removed_Action1No<span class="op">$</span>Action1)</a>
<a class="sourceLine" id="cb61-7" title="7"></a>
<a class="sourceLine" id="cb61-8" title="8"><span class="kw">summary</span>(removed_Action1No<span class="op">$</span>Action1)</a></code></pre></div>
<pre><code>## Unspecified         Yes 
##         678         172</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" title="1"><span class="co">#removed the empty No level since that observation was dropped</span></a>
<a class="sourceLine" id="cb63-2" title="2"></a>
<a class="sourceLine" id="cb63-3" title="3"><span class="kw">summary</span>(removed_Action1No<span class="op">$</span>Country)</a></code></pre></div>
<pre><code>##       Japan    Multiple       Other Unspecified         USA 
##         316          16         409          11          98</code></pre>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" title="1">fixMultiple&lt;-<span class="st"> </span>removed_Action1No <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb65-2" title="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Country=</span> <span class="kw">fct_recode</span>(Country, <span class="st">&quot;Other&quot;</span> =<span class="st"> &quot;Multiple&quot;</span>))</a>
<a class="sourceLine" id="cb65-3" title="3">         </a>
<a class="sourceLine" id="cb65-4" title="4">fixCountryFactors &lt;-<span class="st"> </span>fixMultiple <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb65-5" title="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Country=</span> <span class="kw">fct_recode</span>(Country, <span class="st">&quot;Other&quot;</span> =<span class="st"> &quot;Unspecified&quot;</span>))</a>
<a class="sourceLine" id="cb65-6" title="6"><span class="co">#changed answers with multiple or unspecified in country to other. There is surely a cleaner way to do this.</span></a>
<a class="sourceLine" id="cb65-7" title="7"></a>
<a class="sourceLine" id="cb65-8" title="8"><span class="kw">summary</span>(fixCountryFactors<span class="op">$</span>Country)</a></code></pre></div>
<pre><code>## Japan Other   USA 
##   316   436    98</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1"><span class="kw">summary</span>(fixCountryFactors<span class="op">$</span>Trans1)</a></code></pre></div>
<pre><code>##    Environmental        Foodborne Person to Person          Unknown 
##               11              338              119               61 
##      Unspecified       Waterborne 
##              260               61</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" title="1"><span class="co">#move waterborne into environmental</span></a>
<a class="sourceLine" id="cb69-2" title="2">mergeWaterborne_environmental &lt;-<span class="st"> </span>fixCountryFactors <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb69-3" title="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Trans1=</span> <span class="kw">fct_recode</span>(Trans1, <span class="st">&quot;Environmental&quot;</span> =<span class="st"> &quot;Waterborne&quot;</span>))</a>
<a class="sourceLine" id="cb69-4" title="4"></a>
<a class="sourceLine" id="cb69-5" title="5"><span class="kw">summary</span>(mergeWaterborne_environmental<span class="op">$</span>Trans1)</a></code></pre></div>
<pre><code>##    Environmental        Foodborne Person to Person          Unknown 
##               72              338              119               61 
##      Unspecified 
##              260</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" title="1"><span class="kw">summary</span>(mergeWaterborne_environmental<span class="op">$</span>season)</a></code></pre></div>
<pre><code>##          Fall Spring Summer Winter 
##     55    135    187    103    370</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" title="1"><span class="co">#factor level that is blank labeled unknown</span></a>
<a class="sourceLine" id="cb73-2" title="2">unknown_season &lt;-<span class="st"> </span>mergeWaterborne_environmental <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb73-3" title="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">season=</span> <span class="kw">fct_recode</span>(season, <span class="st">&quot;Unknown&quot;</span> =<span class="st"> &quot;&quot;</span>))</a>
<a class="sourceLine" id="cb73-4" title="4"><span class="kw">summary</span>(unknown_season<span class="op">$</span>season)</a></code></pre></div>
<pre><code>## Unknown    Fall  Spring  Summer  Winter 
##      55     135     187     103     370</code></pre>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" title="1">gg2c4_first &lt;-<span class="st"> </span>unknown_season <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb75-2" title="2"><span class="st">  </span><span class="kw">select</span>(gg2c4, <span class="kw">everything</span>())</a>
<a class="sourceLine" id="cb75-3" title="3"></a>
<a class="sourceLine" id="cb75-4" title="4"><span class="kw">summary</span>(gg2c4_first)</a></code></pre></div>
<pre><code>##  gg2c4            Action1       CasesAll      Country        Deaths       
##  No :591   Unspecified:678   Min.   :    1   Japan:316   Min.   :0.00000  
##  Yes:259   Yes        :172   1st Qu.:    9   Other:436   1st Qu.:0.00000  
##                              Median :   25   USA  : 98   Median :0.00000  
##                              Mean   :  128               Mean   :0.05176  
##                              3rd Qu.:   65               3rd Qu.:0.00000  
##                              Max.   :32150               Max.   :9.00000  
##     Hemisphere  Hospitalizations      MeanD1           MeanI1       
##  Northern:797   Min.   : 0.0000   Min.   : 0.000   Min.   : 0.0000  
##  Southern: 53   1st Qu.: 0.0000   1st Qu.: 0.000   1st Qu.: 0.0000  
##                 Median : 0.0000   Median : 0.000   Median : 0.0000  
##                 Mean   : 0.5518   Mean   : 1.311   Mean   : 0.7165  
##                 3rd Qu.: 0.0000   3rd Qu.: 0.000   3rd Qu.: 0.0000  
##                 Max.   :99.0000   Max.   :96.000   Max.   :48.0000  
##     MedianD1         MedianI1          OBYear             Path1    
##  Min.   : 0.000   Min.   : 0.000   Min.   :1983   No         :196  
##  1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.:1999   Unspecified:614  
##  Median : 0.000   Median : 0.000   Median :2002   Yes        : 40  
##  Mean   : 2.312   Mean   : 1.722   Mean   :2002                    
##  3rd Qu.: 0.000   3rd Qu.: 0.000   3rd Qu.:2005                    
##  Max.   :72.000   Max.   :65.000   Max.   :2010                    
##     RiskAll            season                 Trans1        Vomit       
##  Min.   :    0.0   Unknown: 55   Environmental   : 72   Min.   :0.0000  
##  1st Qu.:    0.0   Fall   :135   Foodborne       :338   1st Qu.:0.0000  
##  Median :   22.0   Spring :187   Person to Person:119   Median :0.0000  
##  Mean   :  359.8   Summer :103   Unknown         : 61   Mean   :0.4894  
##  3rd Qu.:  112.0   Winter :370   Unspecified     :260   3rd Qu.:1.0000  
##  Max.   :24000.0                                        Max.   :1.0000  
##        Setting   
##  Other     :691  
##  Restaurant:159  
##                  
##                  
##                  
## </code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" title="1"><span class="co">#gg2c4_first</span></a></code></pre></div>
<p>At this step, you should have a dataframe containing 850 observations, and 18 variables: 1 outcome, 9 numeric/integer predictors, and 8 factor variables. There should be no missing values. The outcome, <code>gg2c4</code>, should be in the 1st slot.</p>
</div>
<div id="data-splitting" class="section level1">
<h1>Data splitting</h1>
<p>We could do data splitting again as we did in the previous exercise, to have a final test set. But since you saw how it works in the previous exercise, we skip it here. We use the full data to fit to our models. We’ll still use cross-validation to get a more honest estimate of model performance. For a real data analysis, the choice to keep some for a final test or not is based on your goals. If your focus is predictive performance, you should consider this split. If your focus is inference or exploratory analysis, you might want to skip this.</p>
</div>
<div id="model-fitting" class="section level1">
<h1>Model fitting</h1>
<p>So I had planned to use <code>caret</code> exclusivley for this course. Last time I tried feature/subset selection with <code>caret</code>, I found it buggy and not too well documented. I had hoped this had improved since. Unortunately, I was again not really able to get things to work. So even though I said we likely won’t use the <code>mlr</code> package, I decided that to be able to nicely practice feature/subset selection, we need to do so. It’s not a bad idea to get familiar with that package, at times <code>caret</code> can do things <code>mlr</code> can’t do and the reverse. So knowing how to use both is good. We’ll thus use <code>mlr</code> to do our model fitting in this exercise.</p>
</div>
<div id="parallelization" class="section level1">
<h1>Parallelization</h1>
<p><code>mlr</code> allows you to run things in parallel by using multiple cores (<code>caret</code> does too). For instance, if you do 5x cross-validation 5x repeated, you essentially run a very similar piece of code 25 times. Normally, you would run one at a time. But if you have it on a machine with 25 cores, all those 25 could run at the same time, giving you a speed-up of 25 (or close to, there is usually some overhead related to doing things in parallel). This kind of parallel computing is sometimes called <em>embarassingly parallel</em>, because it’s so embarassingly simple to split the task into parallel streams.</p>
<p>Since doing the subset selection below starts getting slow, and because it’s a good topic to know about, we are going to use parallelization here. <code>mlr</code> uses the package <code>parallelMap</code> for this. All you need to do is specify the number of cores/processors you want to use and start the parallel system, and <code>mlr</code> then automatically does things in parallel if feasible.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" title="1"><span class="co">#set the number of cores you want to use. Note that this is actually the number of &#39;logical processors&#39;, which is 2x the number of cores. On a windows machine, the number of cores/logical processors can be found in the task manager. You should only set it to what your computer has (or less). So if your computer has say 4 or 6 cores, you can set it to that, or some lower number. Setting it to a number higher than the cores your computer has doesn&#39;t further speed up things, in fact it slows things down.</span></a>
<a class="sourceLine" id="cb78-2" title="2">ncpu=<span class="dv">4</span>;</a>
<a class="sourceLine" id="cb78-3" title="3"><span class="co">#if you don&#39;t want to run things in parallel, or don&#39;t have multiple cores (unlikely nowadays), </span></a>
<a class="sourceLine" id="cb78-4" title="4"><span class="co">#just comment out the line below.</span></a>
<a class="sourceLine" id="cb78-5" title="5"><span class="kw">parallelStartSocket</span>(ncpu, <span class="dt">show.info=</span><span class="ot">FALSE</span>) </a></code></pre></div>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>Some setup settings that are used in various code chunks below.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" title="1"><span class="co">#These are the backbone :)</span></a>
<a class="sourceLine" id="cb79-2" title="2">outcome &lt;-<span class="st"> </span>gg2c4_first<span class="op">$</span>gg2c4 </a>
<a class="sourceLine" id="cb79-3" title="3">outcomename =<span class="st"> &quot;gg2c4&quot;</span></a>
<a class="sourceLine" id="cb79-4" title="4">predictors &lt;-<span class="st"> </span>gg2c4_first[,<span class="op">-</span><span class="dv">1</span>]</a>
<a class="sourceLine" id="cb79-5" title="5">npred=<span class="kw">ncol</span>(predictors)</a>
<a class="sourceLine" id="cb79-6" title="6"><span class="co">#set sampling method for performance evaluation</span></a>
<a class="sourceLine" id="cb79-7" title="7"><span class="co">#here, we use 5-fold cross-validation, 5-times repeated</span></a>
<a class="sourceLine" id="cb79-8" title="8">sampling_choice =<span class="st"> </span><span class="kw">makeResampleDesc</span>(<span class="st">&quot;RepCV&quot;</span>, <span class="dt">reps =</span> <span class="dv">5</span>, <span class="dt">folds =</span> <span class="dv">5</span>)</a></code></pre></div>
<div id="a-null-model" class="section level2">
<h2>A null model</h2>
<p>To define a null model, we need to determine what performance measure we want to track. As mentioned in the course materials, there are different performance measures. Accuracy or misclassification error is simple, it just counts the number of times the model got it right/wrong. We’ll start with that one, and then try another one later. <code>mlr</code> allows for a lot of different performance measures for both categorical and continuous outcomes, see <a href="https://mlr.mlr-org.com/articles/tutorial/performance.html">here</a> and <a href="https://mlr.mlr-org.com/articles/tutorial/measures.html">here</a>.</p>
<p>For accuracy, the simplest null model always predicts the most frequent category. We can use that as baseline performance.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" title="1"><span class="co">#write code that computes accuracy for a null model</span></a>
<a class="sourceLine" id="cb80-2" title="2"><span class="co">#the null model always predicts &quot;No&quot;</span></a>
<a class="sourceLine" id="cb80-3" title="3"></a>
<a class="sourceLine" id="cb80-4" title="4">null_model &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">850</span></a>
<a class="sourceLine" id="cb80-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">850</span>)</a>
<a class="sourceLine" id="cb80-6" title="6">  null_model[i] =<span class="st"> &quot;No&quot;</span></a>
<a class="sourceLine" id="cb80-7" title="7"></a>
<a class="sourceLine" id="cb80-8" title="8"><span class="kw">measureACC</span>(gg2c4_first<span class="op">$</span>gg2c4, null_model)</a></code></pre></div>
<pre><code>## [1] 0.6952941</code></pre>
<p>You should find that the null model has an accuracy of around 0.69.</p>
</div>
<div id="single-predictor-models" class="section level2">
<h2>Single predictor models</h2>
<p>Now let’s consider single predictor models, i.e. we’ll fit the outcome to each predictor one at a time to get an idea of the importance of individual predictors. To evaluate our model performance, we will use cross-validation. Since our outcome is categorical, we’ll use a logistic model.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb82-2" title="2"><span class="co">#set learner/model. this corresponds to a logistic model.</span></a>
<a class="sourceLine" id="cb82-3" title="3"><span class="co">#mlr calls different models different &quot;learners&quot;</span></a>
<a class="sourceLine" id="cb82-4" title="4">learner_name =<span class="st"> &quot;classif.binomial&quot;</span>;</a>
<a class="sourceLine" id="cb82-5" title="5">mylearner =<span class="st"> </span><span class="kw">makeLearner</span>(learner_name, <span class="dt">predict.type =</span> <span class="st">&quot;prob&quot;</span>)</a>
<a class="sourceLine" id="cb82-6" title="6"><span class="co"># this will contain the results</span></a>
<a class="sourceLine" id="cb82-7" title="7">unifmat=<span class="kw">data.frame</span>(<span class="dt">variable =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred), <span class="dt">Accuracy =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred))</a>
<a class="sourceLine" id="cb82-8" title="8"><span class="co"># loop over each predictor, build simple dataset with just outcome and that predictor, fit it to a glm/logistic model</span></a>
<a class="sourceLine" id="cb82-9" title="9"><span class="cf">for</span> (nn <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>npred)</a>
<a class="sourceLine" id="cb82-10" title="10">{</a>
<a class="sourceLine" id="cb82-11" title="11">    unidata =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gg2c4 =</span> outcome, gg2c4_first[,nn<span class="op">+</span><span class="dv">1</span>] )</a>
<a class="sourceLine" id="cb82-12" title="12">    <span class="co">## Generate the task, i.e. define outcome and predictors to be fit</span></a>
<a class="sourceLine" id="cb82-13" title="13">    mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;unianalysis&#39;</span>, <span class="dt">data =</span> unidata, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb82-14" title="14">    model =<span class="st"> </span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>, <span class="dt">measures =</span> acc )</a>
<a class="sourceLine" id="cb82-15" title="15">    unifmat[nn,<span class="dv">1</span>] =<span class="st"> </span><span class="kw">names</span>(predictors)[nn] </a>
<a class="sourceLine" id="cb82-16" title="16">    unifmat[nn,<span class="dv">2</span>] =<span class="st"> </span>model<span class="op">$</span>aggr</a>
<a class="sourceLine" id="cb82-17" title="17">}</a>
<a class="sourceLine" id="cb82-18" title="18"><span class="kw">kable</span>(unifmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">variable</th>
<th align="right">Accuracy</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Action1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">CasesAll</td>
<td align="right">0.6941176</td>
</tr>
<tr class="odd">
<td align="left">Country</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">Deaths</td>
<td align="right">0.6978824</td>
</tr>
<tr class="odd">
<td align="left">Hemisphere</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">Hospitalizations</td>
<td align="right">0.6936471</td>
</tr>
<tr class="odd">
<td align="left">MeanD1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">MeanI1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="odd">
<td align="left">MedianD1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">MedianI1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="odd">
<td align="left">OBYear</td>
<td align="right">0.6945882</td>
</tr>
<tr class="even">
<td align="left">Path1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="odd">
<td align="left">RiskAll</td>
<td align="right">0.6969412</td>
</tr>
<tr class="even">
<td align="left">season</td>
<td align="right">0.6934118</td>
</tr>
<tr class="odd">
<td align="left">Trans1</td>
<td align="right">0.6952941</td>
</tr>
<tr class="even">
<td align="left">Vomit</td>
<td align="right">0.6952941</td>
</tr>
<tr class="odd">
<td align="left">Setting</td>
<td align="right">0.6952941</td>
</tr>
</tbody>
</table>
<p>So looks like none of the single predictor models have a higher accuracy than the null. Maybe surprising. We’ll get back to that below.</p>
</div>
</div>
<div id="full-model" class="section level1">
<h1>Full model</h1>
<p>Now let’s fit a full logistic model with all predictors.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb83-2" title="2"><span class="co">#do full model with Cross-Validation - to get an idea for the amount of over-fitting a full model does</span></a>
<a class="sourceLine" id="cb83-3" title="3">mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;fullanalysis&#39;</span>, <span class="dt">data =</span> gg2c4_first, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb83-4" title="4">fullmodel =<span class="st"> </span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>, , <span class="dt">measures =</span> acc)</a>
<a class="sourceLine" id="cb83-5" title="5">ACC_fullmodel =<span class="st"> </span>fullmodel<span class="op">$</span>aggr[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb83-6" title="6"><span class="kw">print</span>(ACC_fullmodel)</a></code></pre></div>
<pre><code>## acc.test.mean 
##     0.7054118</code></pre>
<p>So a very small improvement over the simple models.</p>
<p>Now let’s do subset selection. The code below does it several ways. It does regular forward and backward selection and floating versions of those. It also uses a genetic algorithm for selection. See the <code>mlr</code> website <a href="https://mlr.mlr-org.com/articles/tutorial/feature_selection.html">here</a> for details.</p>
<p>Also note that I included a kind of timer in the code, to see how long things take. That’s a good idea if you run bigger models. You first run a few iterations on maybe a few cores, then you can compute how long it would take if you doubled the iterations, or doubled the cores, etc. This prevents bad surprises of trying to <em>quickly</em> run a piece of code and waiting hours. You should always use short runs to make sure everything works in principle, and only at the end do the real, long “production” runs. Otherwise you might waste hours/days/weeks waiting for results only to realize that you made a basic mistake and you have to do it all over.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" title="1"><span class="co">#&quot;subset selection took 1.932833 minutes&quot; for my laptop on this batch of code below</span></a>
<a class="sourceLine" id="cb85-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) </a>
<a class="sourceLine" id="cb85-3" title="3">tstart=<span class="kw">proc.time</span>(); <span class="co">#capture current CPU time for timing how long things take</span></a>
<a class="sourceLine" id="cb85-4" title="4"><span class="co">#do 2 forms of forward and backward selection, just to compare</span></a>
<a class="sourceLine" id="cb85-5" title="5">select_methods=<span class="kw">c</span>(<span class="st">&quot;sbs&quot;</span>,<span class="st">&quot;sfbs&quot;</span>,<span class="st">&quot;sfs&quot;</span>,<span class="st">&quot;sffs&quot;</span>) </a>
<a class="sourceLine" id="cb85-6" title="6">resmat=<span class="kw">data.frame</span>(<span class="dt">method =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">Accuracy =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">Model =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</a>
<a class="sourceLine" id="cb85-7" title="7">ct=<span class="dv">1</span>;</a>
<a class="sourceLine" id="cb85-8" title="8"><span class="cf">for</span> (select_method <span class="cf">in</span> select_methods) <span class="co">#loop over all stepwise selection methods</span></a>
<a class="sourceLine" id="cb85-9" title="9">{</a>
<a class="sourceLine" id="cb85-10" title="10">  ctrl =<span class="st"> </span><span class="kw">makeFeatSelControlSequential</span>(<span class="dt">method =</span> select_method)</a>
<a class="sourceLine" id="cb85-11" title="11">  <span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with method %s &#39;</span>,select_method))</a>
<a class="sourceLine" id="cb85-12" title="12">  sfeat_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb85-13" title="13">                             <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb85-14" title="14">                             <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb85-15" title="15">                             <span class="dt">control =</span> ctrl, </a>
<a class="sourceLine" id="cb85-16" title="16">                             <span class="dt">show.info =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb85-17" title="17">                             <span class="dt">measures =</span> acc)</a>
<a class="sourceLine" id="cb85-18" title="18">  </a>
<a class="sourceLine" id="cb85-19" title="19">  resmat[ct,<span class="dv">1</span>] =<span class="st"> </span>select_methods[ct]</a>
<a class="sourceLine" id="cb85-20" title="20">  resmat[ct,<span class="dv">2</span>] =<span class="st"> </span>sfeat_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb85-21" title="21">  resmat[ct,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeat_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb85-22" title="22">  ct=ct<span class="op">+</span><span class="dv">1</span>;</a>
<a class="sourceLine" id="cb85-23" title="23">}</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with method sbs &quot;
## [1] &quot;doing subset selection with method sfbs &quot;
## [1] &quot;doing subset selection with method sfs &quot;
## [1] &quot;doing subset selection with method sffs &quot;</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb87-1" title="1"><span class="co"># do feature selection with genetic algorithm</span></a>
<a class="sourceLine" id="cb87-2" title="2">maxit =<span class="st"> </span><span class="dv">100</span> <span class="co">#number of iterations - should be large for &#39;production run&#39;</span></a>
<a class="sourceLine" id="cb87-3" title="3">ctrl_GA =<span class="kw">makeFeatSelControlGA</span>(<span class="dt">maxit =</span> maxit)</a>
<a class="sourceLine" id="cb87-4" title="4"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with genetic algorithm&#39;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with genetic algorithm&quot;</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb89-1" title="1">sfeatga_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb89-2" title="2">                                   <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb89-3" title="3">                                   <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb89-4" title="4">                                   <span class="dt">control =</span> ctrl_GA, </a>
<a class="sourceLine" id="cb89-5" title="5">                                   <span class="dt">show.info =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb89-6" title="6">                                   <span class="dt">measures =</span> acc)</a>
<a class="sourceLine" id="cb89-7" title="7">resmat[<span class="dv">5</span>,<span class="dv">1</span>] =<span class="st"> &quot;GA&quot;</span></a>
<a class="sourceLine" id="cb89-8" title="8">resmat[<span class="dv">5</span>,<span class="dv">2</span>] =<span class="st"> </span>sfeatga_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb89-9" title="9">resmat[<span class="dv">5</span>,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeatga_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb89-10" title="10">runtime.minutes_SS=(<span class="kw">proc.time</span>()<span class="op">-</span>tstart)[[<span class="dv">3</span>]]<span class="op">/</span><span class="dv">60</span>; <span class="co">#total time in minutes the optimization took</span></a>
<a class="sourceLine" id="cb89-11" title="11"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;subset selection took %f minutes&#39;</span>,runtime.minutes_SS));</a></code></pre></div>
<pre><code>## [1] &quot;subset selection took 1.848833 minutes&quot;</code></pre>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb91-1" title="1"><span class="kw">kable</span>(resmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">method</th>
<th align="right">Accuracy</th>
<th align="left">Model</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">sbs</td>
<td align="right">0.7197647</td>
<td align="left">Action1, Country, Deaths, MedianI1, OBYear, RiskAll, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="even">
<td align="left">sfbs</td>
<td align="right">0.7136471</td>
<td align="left">Action1, Country, Deaths, OBYear, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="odd">
<td align="left">sfs</td>
<td align="right">0.6952941</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">sffs</td>
<td align="right">0.6952941</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">GA</td>
<td align="right">0.7228235</td>
<td align="left">Action1, Country, Deaths, MedianI1, OBYear, RiskAll, season, Trans1, Vomit, Setting</td>
</tr>
</tbody>
</table>
<p>So we find that some of the sub-models perform somewhat better. Note that the different methods find different submodels and the forward selection method seems to fail (since none of the single-predictor models is better than the null, it stops there). Also note that the genetic algorithm was only allowed to run for a short time (the number if iterations is small). To get potentially better results, that should be increased to a larger number, e.g. 1000 or more. But this will take a longer time, unless you have many cores you can use when you parallelize.</p>
</div>
<div id="different-performance-measures" class="section level1">
<h1>Different performance measures</h1>
<p>So far, we found that some of the sub-models provide a minor improvement over the simple null or single-predictor models or the full model. However, the improvement is not much.</p>
<p>We could try tweaking further, e.g. by pre-processing the predictors (which is a good idea to try, but we won’t do here) or testing a different model (which we’ll do below).</p>
<p>For now, I want to discuss performance measure a bit more. The problem is that logistic regression, as well as many (though not all) algorithms that are used for classification (predicting categorical outcomes) return probabilities for an outcome to belong to a certain category. Those are then turned into Yes/No predictions (in the case with 2 outcomes like we have here), based on some cut-off. By default, a probability value of 0.5 is chosen. If the model predicts that there is a 0.5 or greater probability of the outcome being “Yes”, it is scored as such in the prediction, otherwise as no. However, this 0.5 threshold is not always good. It could be that we might get a much better model if we use another threshold. There is a technique that scans over different thresholds and evaluates model performance for each threshold. This is generally called <em>receiver operating curve</em>. If this concept is new to you, check out <a href="https://en.wikipedia.org/wiki/Receiver_operating_characteristic">this article on Wikipedia</a> or <a href="https://mlr.mlr-org.com/articles/tutorial/roc_analysis.html">this tutorial on the mlr website</a>.</p>
<p>Using this approach, a model is evaluated based on the Area under the curve (AUC), with an AUC=0.5 meaning a model is not better than chance, and AUC=1 meaning a perfect model. Let’s re-do the analysis above, but replace accuracy with AUC.</p>
</div>
<div id="model-performance-with-auc" class="section level1">
<h1>Model performance with AUC</h1>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" title="1"><span class="co">#write code that computes AUC for a null model</span></a>
<a class="sourceLine" id="cb92-2" title="2"><span class="co">#the null model always predicts &quot;No&quot;</span></a>
<a class="sourceLine" id="cb92-3" title="3"><span class="co">#the auc function in the pROC package is useful</span></a>
<a class="sourceLine" id="cb92-4" title="4"><span class="co">#or you can use any other package/function you want to use to compute AUC/ROC</span></a>
<a class="sourceLine" id="cb92-5" title="5"></a>
<a class="sourceLine" id="cb92-6" title="6"><span class="co">#I went ahead and downloaded, installed, and loaded package with other libraries above for pROC. Let&#39;s try it out.</span></a>
<a class="sourceLine" id="cb92-7" title="7"></a>
<a class="sourceLine" id="cb92-8" title="8">null_numeric &lt;-<span class="st"> </span><span class="kw">numeric</span>()</a>
<a class="sourceLine" id="cb92-9" title="9"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">850</span>)</a>
<a class="sourceLine" id="cb92-10" title="10">  null_numeric[i] =<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb92-11" title="11"></a>
<a class="sourceLine" id="cb92-12" title="12">number_roc &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb92-13" title="13"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">gg2c4=</span> <span class="kw">fct_recode</span>(gg2c4, <span class="st">&quot;1&quot;</span> =<span class="st"> &quot;Yes&quot;</span>, <span class="st">&quot;0&quot;</span> =<span class="st"> &quot;No&quot;</span>))</a>
<a class="sourceLine" id="cb92-14" title="14"></a>
<a class="sourceLine" id="cb92-15" title="15">binary_roc &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(number_roc<span class="op">$</span>gg2c4)</a>
<a class="sourceLine" id="cb92-16" title="16"></a>
<a class="sourceLine" id="cb92-17" title="17">roc_null &lt;-<span class="st"> </span><span class="kw">roc</span>(binary_roc, null_numeric)</a></code></pre></div>
<pre><code>## Setting levels: control = 1, case = 2</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb95-1" title="1">auc_null &lt;-<span class="st"> </span><span class="kw">auc</span>(roc_null)</a>
<a class="sourceLine" id="cb95-2" title="2"><span class="kw">plot</span>(auc_null)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/nullmodel-auc-1.png" width="672" /></p>
<p>You should find that the null model has an AUC of 0.5, as expected for a “no information” model.</p>
<p>Now we run the single-predictor models again.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" title="1"><span class="co">#copy and paste the code from above for single predictor fits, but now switch to AUC instead of accuracy.</span></a>
<a class="sourceLine" id="cb96-2" title="2"><span class="co"># loop over each predictor, build simple dataset with just outcome and that predictor, fit it to a glm/logistic model</span></a>
<a class="sourceLine" id="cb96-3" title="3"></a>
<a class="sourceLine" id="cb96-4" title="4"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb96-5" title="5"></a>
<a class="sourceLine" id="cb96-6" title="6"><span class="co"># I played around a lot in this section to learn the ins and outs of using learners and implementing different algorithms</span></a>
<a class="sourceLine" id="cb96-7" title="7"><span class="co">#outcome &lt;- gg2c4_first$gg2c4 </span></a>
<a class="sourceLine" id="cb96-8" title="8"><span class="co">#outcomename = &quot;gg2c4&quot;</span></a>
<a class="sourceLine" id="cb96-9" title="9"><span class="co">#predictors &lt;- gg2c4_first[,-1]</span></a>
<a class="sourceLine" id="cb96-10" title="10"></a>
<a class="sourceLine" id="cb96-11" title="11"><span class="co">#gg2c4_first</span></a>
<a class="sourceLine" id="cb96-12" title="12"></a>
<a class="sourceLine" id="cb96-13" title="13"><span class="co">#subset_negative &lt;- gg2c4_first %&gt;%</span></a>
<a class="sourceLine" id="cb96-14" title="14"><span class="co">#  mutate(gg2c4= fct_recode(gg2c4, &#39;1&#39; =&quot;No&quot;, &#39;0&#39; = &quot;Yes&quot;))</span></a>
<a class="sourceLine" id="cb96-15" title="15"></a>
<a class="sourceLine" id="cb96-16" title="16">positive_binary_tag &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb96-17" title="17"><span class="st">   </span><span class="kw">mutate</span>(<span class="dt">gg2c4=</span> <span class="kw">fct_recode</span>(gg2c4, <span class="st">&quot;Yes&quot;</span> =<span class="st"> &#39;1&#39;</span>, <span class="st">&quot;No&quot;</span> =<span class="st"> &#39;0&#39;</span>))</a>
<a class="sourceLine" id="cb96-18" title="18"></a>
<a class="sourceLine" id="cb96-19" title="19"></a>
<a class="sourceLine" id="cb96-20" title="20">subset_negative &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb96-21" title="21"><span class="st">  </span><span class="kw">filter</span>(gg2c4 <span class="op">==</span><span class="st"> &quot;No&quot;</span>)</a>
<a class="sourceLine" id="cb96-22" title="22"></a>
<a class="sourceLine" id="cb96-23" title="23">subset_positive &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb96-24" title="24"><span class="st">  </span><span class="kw">filter</span>(gg2c4 <span class="op">==</span><span class="st"> &quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb96-25" title="25"></a>
<a class="sourceLine" id="cb96-26" title="26"><span class="co">#subset_negative</span></a>
<a class="sourceLine" id="cb96-27" title="27"><span class="co">#subset_positive</span></a>
<a class="sourceLine" id="cb96-28" title="28"></a>
<a class="sourceLine" id="cb96-29" title="29"><span class="co">#predictors</span></a>
<a class="sourceLine" id="cb96-30" title="30"></a>
<a class="sourceLine" id="cb96-31" title="31">roc &lt;-<span class="st"> </span>(gg2c4_first<span class="op">$</span>gg2c4)</a>
<a class="sourceLine" id="cb96-32" title="32">roc &lt;-<span class="st"> </span><span class="kw">as.list</span>(roc)</a>
<a class="sourceLine" id="cb96-33" title="33"><span class="co">#roc</span></a>
<a class="sourceLine" id="cb96-34" title="34"></a>
<a class="sourceLine" id="cb96-35" title="35"><span class="co">#auc &lt;- measureAUC(gg2c4_first$gg2c4, outcome, subset_negative$gg2c4, subset_positive$gg2c4)</span></a>
<a class="sourceLine" id="cb96-36" title="36"><span class="co">#auc</span></a>
<a class="sourceLine" id="cb96-37" title="37"><span class="co">#plot(auc)</span></a>
<a class="sourceLine" id="cb96-38" title="38"><span class="co">#auc &lt;- auc(gg2c4_first, positive_binary_tag)</span></a>
<a class="sourceLine" id="cb96-39" title="39"><span class="co">#auc</span></a>
<a class="sourceLine" id="cb96-40" title="40"></a>
<a class="sourceLine" id="cb96-41" title="41"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb96-42" title="42"><span class="co">#set learner/model. this corresponds to a logistic model.</span></a>
<a class="sourceLine" id="cb96-43" title="43"><span class="co">#mlr calls different models different &quot;learners&quot;</span></a>
<a class="sourceLine" id="cb96-44" title="44">learner_name =<span class="st"> &quot;classif.logreg&quot;</span>;</a>
<a class="sourceLine" id="cb96-45" title="45">mylearner =<span class="st"> </span><span class="kw">makeLearner</span>(learner_name, <span class="dt">predict.type =</span> <span class="st">&quot;prob&quot;</span>)</a>
<a class="sourceLine" id="cb96-46" title="46"><span class="co"># this will contain the results</span></a>
<a class="sourceLine" id="cb96-47" title="47">unifmat=<span class="kw">data.frame</span>(<span class="dt">variable =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred), <span class="dt">AUC =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred))</a>
<a class="sourceLine" id="cb96-48" title="48">    </a>
<a class="sourceLine" id="cb96-49" title="49">      </a>
<a class="sourceLine" id="cb96-50" title="50"><span class="cf">for</span> (nn <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>npred)</a>
<a class="sourceLine" id="cb96-51" title="51">{</a>
<a class="sourceLine" id="cb96-52" title="52">    auc &lt;-<span class="st"> </span><span class="kw">measureAUC</span>(gg2c4_first[,nn<span class="op">+</span><span class="dv">1</span>], outcome, subset_negative<span class="op">$</span>gg2c4, subset_positive<span class="op">$</span>gg2c4)</a>
<a class="sourceLine" id="cb96-53" title="53">    auc</a>
<a class="sourceLine" id="cb96-54" title="54">    unidata =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gg2c4 =</span> outcome, gg2c4_first[,nn<span class="op">+</span><span class="dv">1</span>] )</a>
<a class="sourceLine" id="cb96-55" title="55">    <span class="co">#auc &lt;- auc(roc(gg2c4_first$gg2c4, positive_binary_tag))</span></a>
<a class="sourceLine" id="cb96-56" title="56">    <span class="co">## Generate the task, i.e. define outcome and predictors to be fit</span></a>
<a class="sourceLine" id="cb96-57" title="57">    mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;unianalysis&#39;</span>, <span class="dt">data =</span> unidata, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb96-58" title="58">    model =<span class="st"> </span>mlr<span class="op">::</span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb96-59" title="59">    unifmat[nn,<span class="dv">1</span>] =<span class="st"> </span><span class="kw">names</span>(predictors)[nn]</a>
<a class="sourceLine" id="cb96-60" title="60">    unifmat[nn,<span class="dv">2</span>] =<span class="st"> </span>model<span class="op">$</span>aggr</a>
<a class="sourceLine" id="cb96-61" title="61">}</a>
<a class="sourceLine" id="cb96-62" title="62"><span class="kw">kable</span>(unifmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">variable</th>
<th align="right">AUC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Action1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">CasesAll</td>
<td align="right">0.3058824</td>
</tr>
<tr class="odd">
<td align="left">Country</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Deaths</td>
<td align="right">0.3021176</td>
</tr>
<tr class="odd">
<td align="left">Hemisphere</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Hospitalizations</td>
<td align="right">0.3063529</td>
</tr>
<tr class="odd">
<td align="left">MeanD1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">MeanI1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">MedianD1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">MedianI1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">OBYear</td>
<td align="right">0.3054118</td>
</tr>
<tr class="even">
<td align="left">Path1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">RiskAll</td>
<td align="right">0.3030588</td>
</tr>
<tr class="even">
<td align="left">season</td>
<td align="right">0.3065882</td>
</tr>
<tr class="odd">
<td align="left">Trans1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Vomit</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">Setting</td>
<td align="right">0.3047059</td>
</tr>
</tbody>
</table>
<p>Looks like there are a few single-predictor models with better performance than the null model.</p>
<p>Now let’s again fit a full model with all predictors.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb97-1" title="1"><span class="co">#copy and paste the code from above for the full model fit, but now switch to AUC instead of accuracy.</span></a>
<a class="sourceLine" id="cb97-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb97-3" title="3"><span class="co">#do full model with Cross-Validation - to get an idea for the amount of over-fitting a full model does</span></a>
<a class="sourceLine" id="cb97-4" title="4"></a>
<a class="sourceLine" id="cb97-5" title="5">mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;fullanalysis&#39;</span>, <span class="dt">data =</span> gg2c4_first, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb97-6" title="6">fullmodel =<span class="st"> </span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb97-7" title="7">AUC_fullmodel =<span class="st"> </span>fullmodel<span class="op">$</span>aggr[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb97-8" title="8"><span class="kw">print</span>(AUC_fullmodel)</a></code></pre></div>
<pre><code>## mmce.test.mean 
##      0.2945882</code></pre>
<p>This is better than any single-predictor models. Let’s see if a sub-model can do even better.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb99-1" title="1"><span class="co">#copy and paste the code from above for the subset selection, but now switch to AUC instead of accuracy.</span></a>
<a class="sourceLine" id="cb99-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) </a>
<a class="sourceLine" id="cb99-3" title="3">tstart=<span class="kw">proc.time</span>(); <span class="co">#capture current CPU time for timing how long things take</span></a>
<a class="sourceLine" id="cb99-4" title="4"><span class="co">#do 2 forms of forward and backward selection, just to compare</span></a>
<a class="sourceLine" id="cb99-5" title="5">select_methods=<span class="kw">c</span>(<span class="st">&quot;sbs&quot;</span>,<span class="st">&quot;sfbs&quot;</span>,<span class="st">&quot;sfs&quot;</span>,<span class="st">&quot;sffs&quot;</span>) </a>
<a class="sourceLine" id="cb99-6" title="6">resmat=<span class="kw">data.frame</span>(<span class="dt">method =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">AUC =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">Model =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</a>
<a class="sourceLine" id="cb99-7" title="7">ct=<span class="dv">1</span>;</a>
<a class="sourceLine" id="cb99-8" title="8"></a>
<a class="sourceLine" id="cb99-9" title="9"><span class="cf">for</span> (select_method <span class="cf">in</span> select_methods) <span class="co">#loop over all stepwise selection methods</span></a>
<a class="sourceLine" id="cb99-10" title="10">{</a>
<a class="sourceLine" id="cb99-11" title="11">  ctrl =<span class="st"> </span><span class="kw">makeFeatSelControlSequential</span>(<span class="dt">method =</span> select_method)</a>
<a class="sourceLine" id="cb99-12" title="12">  <span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with method %s &#39;</span>,select_method))</a>
<a class="sourceLine" id="cb99-13" title="13">  sfeat_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb99-14" title="14">                             <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb99-15" title="15">                             <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb99-16" title="16">                             <span class="dt">control =</span> ctrl, </a>
<a class="sourceLine" id="cb99-17" title="17">                             <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb99-18" title="18">  </a>
<a class="sourceLine" id="cb99-19" title="19">  resmat[ct,<span class="dv">1</span>] =<span class="st"> </span>select_methods[ct]</a>
<a class="sourceLine" id="cb99-20" title="20">  resmat[ct,<span class="dv">2</span>] =<span class="st"> </span>sfeat_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb99-21" title="21">  resmat[ct,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeat_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb99-22" title="22">  ct=ct<span class="op">+</span><span class="dv">1</span>;</a>
<a class="sourceLine" id="cb99-23" title="23">}</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with method sbs &quot;
## [1] &quot;doing subset selection with method sfbs &quot;
## [1] &quot;doing subset selection with method sfs &quot;
## [1] &quot;doing subset selection with method sffs &quot;</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" title="1"><span class="co"># do feature selection with genetic algorithm</span></a>
<a class="sourceLine" id="cb101-2" title="2">maxit =<span class="st"> </span><span class="dv">100</span> <span class="co">#number of iterations - should be large for &#39;production run&#39;</span></a>
<a class="sourceLine" id="cb101-3" title="3">ctrl_GA =<span class="kw">makeFeatSelControlGA</span>(<span class="dt">maxit =</span> maxit)</a>
<a class="sourceLine" id="cb101-4" title="4"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with genetic algorithm&#39;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with genetic algorithm&quot;</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" title="1">sfeatga_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb103-2" title="2">                                   <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb103-3" title="3">                                   <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb103-4" title="4">                                   <span class="dt">control =</span> ctrl_GA, </a>
<a class="sourceLine" id="cb103-5" title="5">                                   <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb103-6" title="6">resmat[<span class="dv">5</span>,<span class="dv">1</span>] =<span class="st"> &quot;GA&quot;</span></a>
<a class="sourceLine" id="cb103-7" title="7">resmat[<span class="dv">5</span>,<span class="dv">2</span>] =<span class="st"> </span>sfeatga_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb103-8" title="8">resmat[<span class="dv">5</span>,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeatga_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb103-9" title="9">runtime.minutes_SS=(<span class="kw">proc.time</span>()<span class="op">-</span>tstart)[[<span class="dv">3</span>]]<span class="op">/</span><span class="dv">60</span>; <span class="co">#total time in minutes the optimization took</span></a>
<a class="sourceLine" id="cb103-10" title="10"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;subset selection took %f minutes&#39;</span>,runtime.minutes_SS));</a></code></pre></div>
<pre><code>## [1] &quot;subset selection took 1.879667 minutes&quot;</code></pre>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" title="1"><span class="kw">kable</span>(resmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">method</th>
<th align="right">AUC</th>
<th align="left">Model</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">sbs</td>
<td align="right">0.2802353</td>
<td align="left">Action1, Country, Deaths, MedianI1, OBYear, RiskAll, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="even">
<td align="left">sfbs</td>
<td align="right">0.2863529</td>
<td align="left">Action1, Country, Deaths, OBYear, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="odd">
<td align="left">sfs</td>
<td align="right">0.3047059</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">sffs</td>
<td align="right">0.3047059</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">GA</td>
<td align="right">0.2771765</td>
<td align="left">Action1, Country, Deaths, MedianI1, OBYear, RiskAll, season, Trans1, Vomit, Setting</td>
</tr>
</tbody>
</table>
<p>Looks like we get some sub-models that are slightly better than the full model, at least as evaluated using the (cross-validated) AUC measure.</p>
</div>
<div id="more-on-rocauc" class="section level1">
<h1>More on ROC/AUC</h1>
<p>While this is a common measure and useful to evaluate how models perform at different cut-off levels, one needs to keep in mind that this doesn’t really measure the performance of a single model, but instead the combined performance of all models with different cut-offs. In practice, if we want to use a model, we have to pick one cut-off. For that purpose, one can look at the whole ROC curve and one then usually chooses the model in the “top-left” corner of the curve, which gives the best overall performance with a good balance between FP and TP. Let’s plot that ROC curve for the model found by the GA.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb106-1" title="1"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb106-2" title="2">d2 &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(gg2c4, sfeatga_res<span class="op">$</span>x )</a>
<a class="sourceLine" id="cb106-3" title="3"><span class="co">#d2</span></a>
<a class="sourceLine" id="cb106-4" title="4">mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;rocanalysis&#39;</span>, <span class="dt">data =</span> d2, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb106-5" title="5">mytask</a></code></pre></div>
<pre><code>## Supervised task: rocanalysis
## Type: classif
## Target: gg2c4
## Observations: 850
## Features:
##    numerics     factors     ordered functionals 
##           5           5           0           0 
## Missings: FALSE
## Has weights: FALSE
## Has blocking: FALSE
## Has coordinates: FALSE
## Classes: 2
##  No Yes 
## 591 259 
## Positive class: Yes</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" title="1">mylearner</a></code></pre></div>
<pre><code>## Learner classif.logreg from package stats
## Type: classif
## Name: Logistic Regression; Short name: logreg
## Class: classif.logreg
## Properties: twoclass,numerics,factors,prob,weights
## Predict-Type: prob
## Hyperparameters: model=FALSE</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" title="1">log_mod =<span class="st"> </span><span class="kw">train</span>(mylearner, mytask)</a>
<a class="sourceLine" id="cb110-2" title="2">log_pred =<span class="st"> </span><span class="kw">predict</span>(log_mod, <span class="dt">task =</span> mytask)</a>
<a class="sourceLine" id="cb110-3" title="3">df =<span class="st"> </span><span class="kw">generateThreshVsPerfData</span>(<span class="kw">list</span>(<span class="dt">logistic =</span> log_pred), <span class="dt">measures =</span> <span class="kw">list</span>(fpr, tpr))</a>
<a class="sourceLine" id="cb110-4" title="4"><span class="kw">plotROCCurves</span>(df)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/roc-plot-1.png" width="672" /></p>
<p>Based on this plot, a FP of around 0.5 an TP of around 0.75 seem to produce a good model. Digging into the <code>df</code> object, you can find the raw data underlying the curve in <code>df$data</code> and will find</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb111-1" title="1"><span class="kw">print</span>(df<span class="op">$</span>data[<span class="dv">29</span><span class="op">:</span><span class="dv">31</span>,])</a></code></pre></div>
<pre><code>##          fpr       tpr threshold
## 29 0.4653130 0.7606178 0.2828283
## 30 0.4399323 0.7413127 0.2929293
## 31 0.4094755 0.7297297 0.3030303</code></pre>
<p>This means for this model, if we were to use it we might want to set a threshold of around 0.3, i.e. everything above that probability would be predicted as positive/yes.</p>
</div>
<div id="trying-a-different-model" class="section level1">
<h1>Trying a different model</h1>
<p>Let’s see if we can find another model that might perform even better. Let’s look at a Linear Discriminant Analysis (LDA) model. See e.g. chapter 4 of ISLR for more on that model. We again consider AUC as our measure. The null-model performance is the same, we need to re-do the remainder.</p>
<p>Let’s re-run the single model, followed by the full model and subset selection.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb113-1" title="1"><span class="co">#copy and paste the code from above for single predictor fits. Switch the learner from a the logistic model to a LDA model. Check the mlr website to figure out the name for that learner. (see &#39;integrated learner&#39; section)</span></a>
<a class="sourceLine" id="cb113-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb113-3" title="3"><span class="co">#set learner/model. this corresponds to a logistic model.</span></a>
<a class="sourceLine" id="cb113-4" title="4"><span class="co">#mlr calls different models different &quot;learners&quot;</span></a>
<a class="sourceLine" id="cb113-5" title="5"></a>
<a class="sourceLine" id="cb113-6" title="6"><span class="co">#This learner is for Linear Discreate Analaysis (LDA)</span></a>
<a class="sourceLine" id="cb113-7" title="7">learner_name =<span class="st"> &quot;classif.lda&quot;</span>;</a>
<a class="sourceLine" id="cb113-8" title="8">mylearner =<span class="st"> </span><span class="kw">makeLearner</span>(learner_name, <span class="dt">predict.type =</span> <span class="st">&quot;prob&quot;</span>)</a>
<a class="sourceLine" id="cb113-9" title="9"><span class="co"># this will contain the results</span></a>
<a class="sourceLine" id="cb113-10" title="10">unifmat=<span class="kw">data.frame</span>(<span class="dt">variable =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred), <span class="dt">AUC =</span> <span class="kw">rep</span>(<span class="dv">0</span>,npred))</a>
<a class="sourceLine" id="cb113-11" title="11">      </a>
<a class="sourceLine" id="cb113-12" title="12">      </a>
<a class="sourceLine" id="cb113-13" title="13"><span class="cf">for</span> (nn <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>npred)</a>
<a class="sourceLine" id="cb113-14" title="14">{</a>
<a class="sourceLine" id="cb113-15" title="15">    <span class="co">#auc &lt;- measureAUC(gg2c4_first[,nn+1], outcome, subset_negative$gg2c4, subset_positive$gg2c4)</span></a>
<a class="sourceLine" id="cb113-16" title="16">    unidata =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gg2c4 =</span> outcome, gg2c4_first[,nn<span class="op">+</span><span class="dv">1</span>] )</a>
<a class="sourceLine" id="cb113-17" title="17">    <span class="co">#auc &lt;- auc(roc(gg2c4_first$gg2c4, positive_binary_tag))</span></a>
<a class="sourceLine" id="cb113-18" title="18">    <span class="co">## Generate the task, i.e. define outcome and predictors to be fit</span></a>
<a class="sourceLine" id="cb113-19" title="19">    mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;unianalysis&#39;</span>, <span class="dt">data =</span> unidata, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb113-20" title="20">    model =<span class="st"> </span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb113-21" title="21">    unifmat[nn,<span class="dv">1</span>] =<span class="st"> </span><span class="kw">names</span>(predictors)[nn] </a>
<a class="sourceLine" id="cb113-22" title="22">    unifmat[nn,<span class="dv">2</span>] =<span class="st"> </span>model<span class="op">$</span>aggr</a>
<a class="sourceLine" id="cb113-23" title="23">}</a>
<a class="sourceLine" id="cb113-24" title="24"><span class="kw">kable</span>(unifmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">variable</th>
<th align="right">AUC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Action1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">CasesAll</td>
<td align="right">0.3058824</td>
</tr>
<tr class="odd">
<td align="left">Country</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Deaths</td>
<td align="right">0.3023529</td>
</tr>
<tr class="odd">
<td align="left">Hemisphere</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Hospitalizations</td>
<td align="right">0.3063529</td>
</tr>
<tr class="odd">
<td align="left">MeanD1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">MeanI1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">MedianD1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">MedianI1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">OBYear</td>
<td align="right">0.3051765</td>
</tr>
<tr class="even">
<td align="left">Path1</td>
<td align="right">0.3070588</td>
</tr>
<tr class="odd">
<td align="left">RiskAll</td>
<td align="right">0.3028235</td>
</tr>
<tr class="even">
<td align="left">season</td>
<td align="right">0.3065882</td>
</tr>
<tr class="odd">
<td align="left">Trans1</td>
<td align="right">0.3047059</td>
</tr>
<tr class="even">
<td align="left">Vomit</td>
<td align="right">0.3047059</td>
</tr>
<tr class="odd">
<td align="left">Setting</td>
<td align="right">0.3047059</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" title="1"><span class="co">#copy and paste the code from above for the full model fit, but now for LDA. Since you already switched the learner/model above, no further adjustment should be needed</span></a>
<a class="sourceLine" id="cb114-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb114-3" title="3"><span class="co">#do full model with Cross-Validation - to get an idea for the amount of over-fitting a full model does</span></a>
<a class="sourceLine" id="cb114-4" title="4"></a>
<a class="sourceLine" id="cb114-5" title="5">mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;fullanalysis&#39;</span>, <span class="dt">data =</span> gg2c4_first, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb114-6" title="6">fullmodel =<span class="st"> </span><span class="kw">resample</span>(mylearner, <span class="dt">task =</span> mytask, <span class="dt">resampling =</span> sampling_choice, <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb114-7" title="7">AUC_fullmodel =<span class="st"> </span>fullmodel<span class="op">$</span>aggr[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb114-8" title="8"><span class="kw">print</span>(AUC_fullmodel)</a></code></pre></div>
<pre><code>## mmce.test.mean 
##      0.2898824</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" title="1"><span class="co">#copy and paste the code from above for subset selection, now for LDA. Since you already switched the learner/model above, no further adjustment should be needed</span></a>
<a class="sourceLine" id="cb116-2" title="2"><span class="kw">set.seed</span>(<span class="dv">1111</span>) </a>
<a class="sourceLine" id="cb116-3" title="3">tstart=<span class="kw">proc.time</span>(); <span class="co">#capture current CPU time for timing how long things take</span></a>
<a class="sourceLine" id="cb116-4" title="4"><span class="co">#do 2 forms of forward and backward selection, just to compare</span></a>
<a class="sourceLine" id="cb116-5" title="5">select_methods=<span class="kw">c</span>(<span class="st">&quot;sbs&quot;</span>,<span class="st">&quot;sfbs&quot;</span>,<span class="st">&quot;sfs&quot;</span>,<span class="st">&quot;sffs&quot;</span>) </a>
<a class="sourceLine" id="cb116-6" title="6">resmat=<span class="kw">data.frame</span>(<span class="dt">method =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">AUC =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>), <span class="dt">Model =</span> <span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</a>
<a class="sourceLine" id="cb116-7" title="7">ct=<span class="dv">1</span>;</a>
<a class="sourceLine" id="cb116-8" title="8"></a>
<a class="sourceLine" id="cb116-9" title="9"><span class="cf">for</span> (select_method <span class="cf">in</span> select_methods) <span class="co">#loop over all stepwise selection methods</span></a>
<a class="sourceLine" id="cb116-10" title="10">{</a>
<a class="sourceLine" id="cb116-11" title="11">  ctrl =<span class="st"> </span><span class="kw">makeFeatSelControlSequential</span>(<span class="dt">method =</span> select_method)</a>
<a class="sourceLine" id="cb116-12" title="12">  <span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with method %s &#39;</span>,select_method))</a>
<a class="sourceLine" id="cb116-13" title="13">  sfeat_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb116-14" title="14">                             <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb116-15" title="15">                             <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb116-16" title="16">                             <span class="dt">control =</span> ctrl, </a>
<a class="sourceLine" id="cb116-17" title="17">                             <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb116-18" title="18">  </a>
<a class="sourceLine" id="cb116-19" title="19">  resmat[ct,<span class="dv">1</span>] =<span class="st"> </span>select_methods[ct]</a>
<a class="sourceLine" id="cb116-20" title="20">  resmat[ct,<span class="dv">2</span>] =<span class="st"> </span>sfeat_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb116-21" title="21">  resmat[ct,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeat_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb116-22" title="22">  ct=ct<span class="op">+</span><span class="dv">1</span>;</a>
<a class="sourceLine" id="cb116-23" title="23">}</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with method sbs &quot;
## [1] &quot;doing subset selection with method sfbs &quot;
## [1] &quot;doing subset selection with method sfs &quot;
## [1] &quot;doing subset selection with method sffs &quot;</code></pre>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb118-1" title="1"><span class="co"># do feature selection with genetic algorithm</span></a>
<a class="sourceLine" id="cb118-2" title="2">maxit =<span class="st"> </span><span class="dv">100</span> <span class="co">#number of iterations - should be large for &#39;production run&#39;</span></a>
<a class="sourceLine" id="cb118-3" title="3">ctrl_GA =<span class="kw">makeFeatSelControlGA</span>(<span class="dt">maxit =</span> maxit)</a>
<a class="sourceLine" id="cb118-4" title="4"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;doing subset selection with genetic algorithm&#39;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;doing subset selection with genetic algorithm&quot;</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb120-1" title="1">sfeatga_res =<span class="st"> </span><span class="kw">selectFeatures</span>(<span class="dt">learner =</span> mylearner, </a>
<a class="sourceLine" id="cb120-2" title="2">                                   <span class="dt">task =</span> mytask, </a>
<a class="sourceLine" id="cb120-3" title="3">                                   <span class="dt">resampling =</span> sampling_choice, </a>
<a class="sourceLine" id="cb120-4" title="4">                                   <span class="dt">control =</span> ctrl_GA, </a>
<a class="sourceLine" id="cb120-5" title="5">                                   <span class="dt">show.info =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb120-6" title="6">resmat[<span class="dv">5</span>,<span class="dv">1</span>] =<span class="st"> &quot;GA&quot;</span></a>
<a class="sourceLine" id="cb120-7" title="7">resmat[<span class="dv">5</span>,<span class="dv">2</span>] =<span class="st"> </span>sfeatga_res<span class="op">$</span>y</a>
<a class="sourceLine" id="cb120-8" title="8">resmat[<span class="dv">5</span>,<span class="dv">3</span>] =<span class="st"> </span><span class="kw">paste</span>(<span class="kw">as.vector</span>(sfeatga_res<span class="op">$</span>x), <span class="dt">collapse=</span> <span class="st">&#39;, &#39;</span>)</a>
<a class="sourceLine" id="cb120-9" title="9">runtime.minutes_SS=(<span class="kw">proc.time</span>()<span class="op">-</span>tstart)[[<span class="dv">3</span>]]<span class="op">/</span><span class="dv">60</span>; <span class="co">#total time in minutes the optimization took</span></a>
<a class="sourceLine" id="cb120-10" title="10"><span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">&#39;subset selection took %f minutes&#39;</span>,runtime.minutes_SS));</a></code></pre></div>
<pre><code>## [1] &quot;subset selection took 1.488500 minutes&quot;</code></pre>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb122-1" title="1"><span class="kw">kable</span>(resmat)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">method</th>
<th align="right">AUC</th>
<th align="left">Model</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">sbs</td>
<td align="right">0.2821176</td>
<td align="left">Action1, Country, Deaths, Hemisphere, OBYear, Path1, RiskAll, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="even">
<td align="left">sfbs</td>
<td align="right">0.2811765</td>
<td align="left">Action1, Country, Deaths, OBYear, Path1, season, Trans1, Vomit, Setting</td>
</tr>
<tr class="odd">
<td align="left">sfs</td>
<td align="right">0.3047059</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">sffs</td>
<td align="right">0.3047059</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">GA</td>
<td align="right">0.2863529</td>
<td align="left">Action1, Country, Deaths, Hemisphere, RiskAll, Vomit, Setting</td>
</tr>
</tbody>
</table>
<p>It Looks like we again get some sub-models that are slightly better than the full model, at least as evaluated using the (cross-validated) AUC measure. There seems to be a bit, but not much difference in performance to the logistic model. We can take the GA model again (because it’s conveniently at the end, even if it’s not the best) and look at its ROC and compare the curves for the LDA and logistic models.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" title="1"><span class="co">#copy the code from roc-plot above and re-do train/predict but now save it as lda_mod and lda_pred</span></a>
<a class="sourceLine" id="cb123-2" title="2"><span class="co">#then use generateThreshVsPerfData() with both lda_pred and log_pred (computed above) to create a structure that contains</span></a>
<a class="sourceLine" id="cb123-3" title="3"><span class="co">#best model curves for both logistic and LDA, then plot the curves.</span></a>
<a class="sourceLine" id="cb123-4" title="4"></a>
<a class="sourceLine" id="cb123-5" title="5"><span class="kw">set.seed</span>(<span class="dv">1111</span>) <span class="co">#makes each code block reproducible</span></a>
<a class="sourceLine" id="cb123-6" title="6">d2 &lt;-<span class="st"> </span>gg2c4_first <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(gg2c4, sfeatga_res<span class="op">$</span>x )</a>
<a class="sourceLine" id="cb123-7" title="7"></a>
<a class="sourceLine" id="cb123-8" title="8">mytask =<span class="st"> </span><span class="kw">makeClassifTask</span>(<span class="dt">id=</span><span class="st">&#39;rocanalysis&#39;</span>, <span class="dt">data =</span> d2, <span class="dt">target =</span> outcomename, <span class="dt">positive =</span> <span class="st">&quot;Yes&quot;</span>)</a>
<a class="sourceLine" id="cb123-9" title="9"></a>
<a class="sourceLine" id="cb123-10" title="10">lda_mod =<span class="st"> </span><span class="kw">train</span>(mylearner, mytask)</a>
<a class="sourceLine" id="cb123-11" title="11">lda_pred =<span class="st"> </span><span class="kw">predict</span>(lda_mod, <span class="dt">task =</span> mytask)</a>
<a class="sourceLine" id="cb123-12" title="12">lda_pred</a></code></pre></div>
<pre><code>## Prediction: 850 observations
## predict.type: prob
## threshold: No=0.50,Yes=0.50
## time: 0.00
##   id truth   prob.No   prob.Yes response
## 1  1    No 0.7825221 0.21747793       No
## 2  2    No 0.7823488 0.21765119       No
## 3  3    No 0.8484739 0.15152612       No
## 4  4    No 0.9116073 0.08839274       No
## 5  5    No 0.8607772 0.13922279       No
## 6  6    No 0.8772361 0.12276394       No
## ... (#rows: 850, #cols: 5)</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" title="1">log_pred</a></code></pre></div>
<pre><code>## Prediction: 850 observations
## predict.type: prob
## threshold: No=0.50,Yes=0.50
## time: 0.00
##   id truth   prob.No   prob.Yes response
## 1  1    No 0.6828058 0.31719421       No
## 2  2    No 0.6826736 0.31732644       No
## 3  3    No 0.8171145 0.18288554       No
## 4  4    No 0.9069865 0.09301353       No
## 5  5    No 0.9617729 0.03822715       No
## 6  6    No 0.8941284 0.10587157       No
## ... (#rows: 850, #cols: 5)</code></pre>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" title="1">df =<span class="st"> </span><span class="kw">generateThreshVsPerfData</span>(<span class="kw">list</span>(<span class="dt">logistic =</span> lda_pred), <span class="dt">measures =</span> <span class="kw">list</span>(fpr, tpr))</a>
<a class="sourceLine" id="cb127-2" title="2"><span class="kw">plotROCCurves</span>(df)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/roc-plot-2-1.png" width="672" /></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" title="1">df =<span class="st"> </span><span class="kw">generateThreshVsPerfData</span>(<span class="kw">list</span>(<span class="dt">logistic =</span> log_pred, <span class="dt">LDA =</span> lda_pred), <span class="dt">measures =</span> <span class="kw">list</span>(fpr, tpr))</a>
<a class="sourceLine" id="cb128-2" title="2"><span class="kw">plotROCCurves</span>(df)</a></code></pre></div>
<p><img src="Variable_Selection_files/figure-html/roc-plot-2-2.png" width="672" /></p>
<p>You should see from the plot that the LDA model performs very similar to the Logistic model.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" title="1"><span class="co">#close down the parallelization started at the beginning to free up resources.</span></a>
<a class="sourceLine" id="cb129-2" title="2"><span class="kw">parallelStop</span>()</a></code></pre></div>
</div>
<div id="wrapping-up" class="section level1">
<h1>Wrapping up</h1>
<p>Based on the above, we could choose a model simply by best performance. However, we likely also still want to look at model prediction uncertainty and do some residual plots and other diagnostics for our chosen model before we finalize our choice. As was the case for <code>caret</code>, <code>mlr</code> also comes with lots of functions that make these - and many other - tasks easier. We’ll leave it at this for now, and might revisit some of those topics in further exercises.</p>
<p>Also, none of the models here are that great. We might want to think more about our initial premise, i.e. looking for associations between virus strain type and other variables and what the scientific rationale is for expecting variations. Here, we just ran the model and looked what we might find. That approach, sometimes called <em>data exploration</em> or <em>data mining</em> or - less charitable <em>fishing expedition</em> is ok, but we need to be careful how we interpret and use the results. Going in with a clear hypothesis is usually better.</p>
</div>
<div id="a-few-more-comments" class="section level1">
<h1>A few more comments</h1>
<p>Things are getting a bit slow now, despite the multiple cores we need minutes to run things. Also, code chunks get larger. At this point, it might be worth considering a structure whereby most of the code lives in one or several <strong>well documented</strong> R scripts, which can be run independently. Those R scripts should save their results, and those results are then loaded here. The advantage is that if one wants to make changes to a small part of the analysis, one can modify and run just that part and update the whole document by re-knitting without having to run all code. For any big/serious analysis, I suggest such a setup that splits R code from RMarkdown files, at least for the computationally intensive parts.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
